<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Home | OrthoFinder</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="Home" />
<meta name="author" content="David Emms, Jonathan Holmes, Laurence Belcher, Yi Liu" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="This is a software developed by the Steven Kelly Lab." />
<meta property="og:description" content="This is a software developed by the Steven Kelly Lab." />
<link rel="canonical" href="http://localhost:4000/OrthoFinder/" />
<meta property="og:url" content="http://localhost:4000/OrthoFinder/" />
<meta property="og:site_name" content="OrthoFinder" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Home" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebSite","author":{"@type":"Person","name":"David Emms, Jonathan Holmes, Laurence Belcher, Yi Liu"},"description":"This is a software developed by the Steven Kelly Lab.","headline":"Home","name":"OrthoFinder","url":"http://localhost:4000/OrthoFinder/"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/OrthoFinder/assets/css/style.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/OrthoFinder/feed.xml" title="OrthoFinder" /><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
</head>
<body><header class="site-header">

  <div class="wrapper"><!-- <a class="site-title" rel="author" href="/OrthoFinder/%20/">OrthoFinder</a> -->
    <a class="site-title" rel="author" href="/OrthoFinder/">OrthoFinder</a><nav class="site-nav">
      <input type="checkbox" id="nav-trigger" class="nav-trigger" />
      <label for="nav-trigger">
        <span class="menu-icon">
          <svg viewBox="0 0 18 15" width="18px" height="15px">
            <path
              d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z" />
          </svg>
        </span>
      </label>
      <div class="trigger">
        <!--<a class="page-link" href="/OrthoFinder/">Home</a><a class="page-link" href="/OrthoFinder/about/">About</a><a class="page-link" href="/OrthoFinder/download/">Download</a><a class="page-link" href="/OrthoFinder/installation/">Installation</a><a class="page-link" href="/OrthoFinder/tutorial/">Tutorial</a><a class="page-link" href="/OrthoFinder/faq/">FAQ</a>--><a class="page-link" href="/OrthoFinder/">Home</a><a class="page-link" href="/OrthoFinder/about/">About</a><a class="page-link" href="/OrthoFinder/download/">Download</a><a class="page-link" href="/OrthoFinder/installation/">Installation</a><a class="page-link" href="/OrthoFinder/tutorial/">Tutorial</a><a class="page-link" href="/OrthoFinder/faq/">FAQ</a>
        <a class="page-link" href="https://github.com/OrthoFinder/OrthoFinder" target="_blank" title="GitHub Profile">GitHub</a>
        
      </div>
    </nav></div>
  <script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js"></script>
</header><main class="page-content" aria-label="Content">
      <div class="wrapper">
        <div class="page">
  <!--<h1 class="page-heading">Home</h1>-->

  <!-- <h2 id="home">Home</h2>

<p align="center">
  <img src="assets/images/workflow.jpg" alt="workflow" width="1000" />
</p>
<p>OrthoFinder identifies orthogroups, infers gene trees for all orthogroups, and analyzes the gene trees to identify the rooted species tree. The method subsequently identifies all gene duplication events in the complete set of gene trees, and analyses them at both gene tree and species tree level. OrthoFinder further analyzes all of this phylogenetic information to identify the complete set of orthologs between all species, and provide extensive comparative genomics statistics.</p>

<h2 id="table-of-contents">Table of contents</h2>
<ul>
  <li><a href="#Installation">Installation</a></li>
  <li><a href="#Simple-Usage">Simple Usage</a></li>
  <li><a href="#Advanced-Usage-Scaling-to-Thousands-of-Species">Advanced Usage - Scaling to Thousands of Species</a></li>
  <li><a href="#Command-line-Options">Command line Options</a></li>
  <li><a href="#Output-files">Output files</a></li>
  <li><a href="#Latest-additions">Latest additions</a></li>
  <li><a href="#Citation">Citation</a></li>
</ul>

<p>A single PDF with all documentation and tutorial is available <a href="https://github.com/OrthoFinder/OrthoFinder/blob/update_readme/assets/docs/of3_v2.pdf">here</a></p>

<h2 id="installation">Installation</h2>

<h3 id="install-in-conda-recommended">Install in conda (recommended)</h3>

<p>The easiest way to install OrthoFinder is using <a href="https://www.machinelearningplus.com/deployment/conda-create-environment-and-everything-you-need-to-know-to-manage-conda-virtual-environment/">conda</a>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>conda create <span class="nt">-n</span> of3_env <span class="nv">python</span><span class="o">=</span>3.10
conda activate of3_env
conda <span class="nb">install </span>orthofinder
</code></pre></div></div>

<p>Alternatively, you could install via github, or download the source code and install locally.</p>

<h3 id="install-via-github">Install via github</h3>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>python3 <span class="nt">-m</span> venv of3_env 
<span class="nb">.</span> of3_env/bin/activate
pip <span class="nb">install </span>git+https://github.com/OrthoFinder/OrthoFinder.git
</code></pre></div></div>

<h3 id="install-locally-from-source-code">Install locally from source code</h3>

<p>The following commands provide three ways to download the source code of OrthoFinder locally into a directory named <code class="language-plaintext highlighter-rouge">OrthoFinder</code>.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Download via git </span>
git clone https://github.com/OrthoFinder/OrthoFinder.git
<span class="c"># or download the zipfile and unzip it into OrthoFinder</span>
<span class="nb">mkdir </span>OrthoFinder <span class="o">&amp;&amp;</span> wget <span class="nt">-qO-</span> https://github.com/OrthoFinder/OrthoFinder/archive/refs/heads/master.zip | funzip | <span class="nb">tar</span> <span class="nt">-x</span> <span class="nt">--strip-components</span><span class="o">=</span>1 <span class="nt">-C</span> OrthoFinder
<span class="c"># or download the tar.gz and unzip it into OrthoFinder</span>
<span class="nb">mkdir </span>OrthoFinder <span class="o">&amp;&amp;</span> wget <span class="nt">-qO-</span> https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.0.1/orthofinder-3.0.1.tar.gz | <span class="nb">tar</span> <span class="nt">-xz</span> <span class="nt">--strip-components</span><span class="o">=</span>1 <span class="nt">-C</span> OrthoFinder
</code></pre></div></div>

<p>Next, you can run the following commands to install OrthoFinder inside the of3_env virtural environment.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cd </span>OrthoFinder
python3 <span class="nt">-m</span> venv of3_env <span class="o">&amp;&amp;</span> <span class="nb">.</span> of3_env/bin/activate
pip <span class="nb">install</span> <span class="nb">.</span>
</code></pre></div></div>

<p>Whether you’ve installed OrthoFinder directly from GitHub or downloaded and set it up locally, the OrthoFinder package will only be available within the <code class="language-plaintext highlighter-rouge">of3_env</code> virtual environment. This avoids potential conflicts with Python dependencies.</p>

<p>To deactivate the virtual environment when you are finished, run:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>deactivate
</code></pre></div></div>

<h3 id="test-your-installation">Test your installation</h3>

<p>Once you have installed OrthoFinder, you can print the help information and version, and test it on the <a href="https://github.com/OrthoFinder/OrthoFinder/tree/master/ExampleData">example data</a>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>orthofinder <span class="nt">--help</span> <span class="c"># Print out help informatioin</span>
orthofinder <span class="nt">--version</span> <span class="c"># Check the version</span>
orthofinder <span class="nt">-f</span> ExampleData <span class="c"># Test OrthoFinder on an example dataset</span>
</code></pre></div></div>

<h3 id="uninstalling">Uninstalling</h3>
<p>To uninstall on conda:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>conda deactivate
conda remove <span class="nt">-n</span> of3_env <span class="nt">--all</span>
</code></pre></div></div>
<p>To remove the virtual environment where OrthoFinder is installed:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>deactivate
<span class="nb">cd</span> ..
<span class="nb">rm</span> <span class="nt">-rf</span> OrthoFinder
</code></pre></div></div>

<h2 id="simple-usage">Simple Usage</h2>

<p>Run OrthoFinder on FASTA format proteomes in <code class="language-plaintext highlighter-rouge">&lt;dir&gt;</code></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">orthofinder</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span> <span class="o">-</span><span class="n">f</span> <span class="o">&lt;</span><span class="nb">dir</span><span class="o">&gt;</span>
</code></pre></div></div>

<p>OrthoFinder requires one FASTA file for each species. Each file should contain the complete set of protein sequences from that species’ genome,  with a single representative sequence for each gene.</p>

<p>If your files have multiple transcript variants for each gene, then we provide a script <code class="language-plaintext highlighter-rouge">primary_transcripts.py</code> to extract the longest variant per gene. This script should be run on your files prior to running OrthoFinder;</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for </span>f <span class="k">in</span> <span class="k">*</span>fa <span class="p">;</span> <span class="k">do </span>python primary_transcript.py <span class="nv">$f</span> <span class="p">;</span> <span class="k">done</span>
</code></pre></div></div>

<h2 id="advanced-usage---scaling-to-thousands-of-species">Advanced Usage - Scaling to Thousands of Species</h2>

<p>If you are analysing &gt;100 species, we reccommend that you use the scalable implementation.</p>

<p>Add the files for 64 species into one directory <code class="language-plaintext highlighter-rouge">&lt;core&gt;</code>
Add the remaining files into another directory <code class="language-plaintext highlighter-rouge">&lt;additional&gt;</code></p>

<p>First, run OrthoFinder on the subset of 64 species</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">orthofinder</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span> <span class="o">-</span><span class="n">f</span> <span class="o">&lt;</span><span class="n">core</span><span class="o">&gt;</span>
</code></pre></div></div>

<p>Then, add the additional species to the results of the core run</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">orthofinder</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span> <span class="o">--</span><span class="n">assign</span> <span class="o">&lt;</span><span class="n">additional</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">core</span> <span class="o">&lt;</span><span class="n">Results_Dir</span><span class="o">&gt;</span>
</code></pre></div></div>

<p>To choose which 64 species to include in the core, aim to capture a broad range of the evolutionary diversity of your species.</p>

<p align="center">
  <img src="assets/images/core_assign_figure.jpg" alt="core_assign_figure" width="800" />
</p>

<p><strong>Note that this alternative way of running OrthoFinder requires that the core species are run using the multiple sequence alignment option. You cannot add additional species to OrthoFinder results that were run with the <code class="language-plaintext highlighter-rouge">-M dendroblast</code> option, which was the default for OrthoFinder2</strong></p>

<h2 id="command-line-options">Command-line options</h2>

<p>Command-line options for OrthoFinder</p>

<p><strong>Adding additional species</strong>
| Parameter | Description                               |
|———–|——————————————-|
| <code class="language-plaintext highlighter-rouge">--assign &lt;dir1&gt; --core &lt;dir2&gt;</code>      | Assign species from <code class="language-plaintext highlighter-rouge">&lt;dir1&gt;</code> to existing orthogroups in <code class="language-plaintext highlighter-rouge">&lt;dir2&gt;</code>.                   |</p>

<p><strong>Method choices</strong>
| Parameter | Description                               | Default   | Options                                                                                     |
|———–|——————————————-|———–|———————————————————————————————|
| <code class="language-plaintext highlighter-rouge">-M</code>      | Method for gene tree inference.           | <code class="language-plaintext highlighter-rouge">msa</code>     | <code class="language-plaintext highlighter-rouge">dendroblast</code>, <code class="language-plaintext highlighter-rouge">msa</code>                                                                        |
| <code class="language-plaintext highlighter-rouge">-S</code>      | Sequence search program                   | <code class="language-plaintext highlighter-rouge">diamond</code> | <code class="language-plaintext highlighter-rouge">blast</code>, <code class="language-plaintext highlighter-rouge">diamond</code>, <code class="language-plaintext highlighter-rouge">diamond_ultra_sens</code>, <code class="language-plaintext highlighter-rouge">blastp</code>, <code class="language-plaintext highlighter-rouge">mmseqs</code>, <code class="language-plaintext highlighter-rouge">blastn</code> |
| <code class="language-plaintext highlighter-rouge">-A</code>      | MSA program, requires <code class="language-plaintext highlighter-rouge">-M msa</code>            | <code class="language-plaintext highlighter-rouge">famsa</code>   | <code class="language-plaintext highlighter-rouge">famsa</code>, <code class="language-plaintext highlighter-rouge">mafft</code>, <code class="language-plaintext highlighter-rouge">muscle</code>,                                                          |
| <code class="language-plaintext highlighter-rouge">-T</code>      | Tree inference method, requires <code class="language-plaintext highlighter-rouge">-M msa</code>  | <code class="language-plaintext highlighter-rouge">fasttree</code>| <code class="language-plaintext highlighter-rouge">fasttree</code>, <code class="language-plaintext highlighter-rouge">fasttree_fastest</code>, <code class="language-plaintext highlighter-rouge">raxml</code>, <code class="language-plaintext highlighter-rouge">iqtree</code>                               |
| <code class="language-plaintext highlighter-rouge">-I</code>      | MCL inflation parameter                   | <code class="language-plaintext highlighter-rouge">1.2</code>     | <code class="language-plaintext highlighter-rouge">1-10</code>                                                                                         |</p>

<p><strong>Input options</strong>
| Parameter | Description                               |
|———–|——————————————-|
| <code class="language-plaintext highlighter-rouge">-d</code>      | Input is DNA sequences.                   |
| <code class="language-plaintext highlighter-rouge">-s</code>      | User-specified rooted species tree.        |</p>

<p><strong>Output options</strong>
| Parameter   | Description                                                                 |
|———–  |—————————————————————————–|
| <code class="language-plaintext highlighter-rouge">-X</code>      | Don’t add species names to sequence IDs.                                    |
| <code class="language-plaintext highlighter-rouge">-n &lt;txt&gt;</code>      | Name to append to the results directory.                                    |
| <code class="language-plaintext highlighter-rouge">-o &lt;txt&gt;</code>      | Specify a non-default results directory.                                    |</p>

<p><strong>Parallel processing options</strong>
| Parameter | Description                                 | Default |
|———–|———————————————|———|
| <code class="language-plaintext highlighter-rouge">-t</code>      | Number of parallel sequence search threads. | <code class="language-plaintext highlighter-rouge">All available</code>    |
| <code class="language-plaintext highlighter-rouge">-a</code>      | Number of parallel analysis threads.        | <code class="language-plaintext highlighter-rouge">16 or t/8 (whichever lower)</code>     |</p>

<p><strong>Workflow stopping options</strong>
| Parameter | Description                                                                 |
|———–|—————————————————————————–|
| <code class="language-plaintext highlighter-rouge">-op</code>     | Stop after preparing input files for BLAST.                                 |</p>

<p><strong>Workflow restart options</strong>
| Parameter  | Description                                                  |
|————|————————————————————–|
| <code class="language-plaintext highlighter-rouge">-b &lt;dir&gt;</code> | Start OrthoFinder from pre-computed BLAST results in <code class="language-plaintext highlighter-rouge">&lt;dir&gt;</code>. |</p>

<p><strong>Other options</strong>
| Parameter        | Description                                                               |
|——————|—————————————————————————|
| <code class="language-plaintext highlighter-rouge">-1</code>             | Only perform one-way sequence search.                                     |
| <code class="language-plaintext highlighter-rouge">-z</code>             | Don’t trim MSAs (columns &gt;= 90% gap, min. alignment length 500).          |
| <code class="language-plaintext highlighter-rouge">-y</code>             | Split paralogous clades below the root of a HOG into separate HOGs.        |
| <code class="language-plaintext highlighter-rouge">-h</code>             | Print this help text.                                                     |
| <code class="language-plaintext highlighter-rouge">-v</code>             | Print version.                                                     |</p>

<h2 id="output-files">Output files</h2>

<p>A standard OrthoFinder run produces a set of files describing the orthogroups, orthologs, gene trees, resolve gene trees, the rooted species tree, gene duplication events, and comparative genomic statistics for the set of species being analysed. These files are located in an intuitive directory structure.</p>

<p>Full details on the output files and directories can be found in the PDF <a href="link">here</a>. The directories that are useful for most users are;</p>

<p><code class="language-plaintext highlighter-rouge">/Orthogroups</code></p>
<ul>
  <li>Orthogroups.tsv is the main orthogroup file. Each row contains the genes belonging to a single orthogroup. The genes from each orthogroup are organized into columns, one per species.</li>
  <li>Orthogroups.txt is a text file with each line showing the genes in a single orthogroup. It differs from Orthogroups.tsv in that it doesn’t show the species which each gene belongs to.</li>
  <li>Orthogroups.GeneCount.tsv is a tab separated text file that contains counts of the number of genes for each species in each orthogroup.</li>
  <li>Orthogroups_SingleCopyOrthologues.txt is a list of orthogroups that contain exactly one gene per species</li>
  <li>Orthogrouops_UnassignedGenes.tsv is a tab separated text file that contains all of the genes that were not assigned to any orthogroup.</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Phylogenetic_Hierarchical_Orthogroups</code></p>
<ul>
  <li>Each file is a phylogenetic hierarchical orthogroup (HOG) for a different node of the species tree</li>
  <li>Each row of a file contain the genes belonging to a single orthogroup</li>
  <li>Each species is represented by a single column</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Orthologues</code></p>
<ul>
  <li>Each species has a sub-directory that in turn contains a file for each pairwise species comparison, listing the orthologs between that species pair.</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Comparative_Genomics_Statistics</code></p>
<ul>
  <li>Files containing summary statistics across all orthogroups, as well as comparisons between each pair of species</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Resolved_Gene_Trees</code></p>
<ul>
  <li>A rooted phylogenetic tree inferred for each orthogroup with 4 or more sequences and resolved using the OrthoFinder hybrid species-overlap/duplication-loss coalescent model.</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Species_Tree</code></p>
<ul>
  <li>SpeciesTree_rooted.txt = A species tree inferred using ASTRAL-Pro.</li>
  <li>SpeciesTree_rooted_node_labels.txt = The same tree, but with nodes labels instead of support values. This labelled version is useful for interpreting and analysing the results of the gene duplication analyses.</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Gene_Duplication_Events</code></p>
<ul>
  <li><code class="language-plaintext highlighter-rouge">Duplications.tsv</code> has a row for each gene duplication event, with information on orthogroup in which it occured, the species that contain the duplicated gene, the node in the species tree on which the gene duplication event occured, and the support score for the gene duplication event.</li>
  <li><code class="language-plaintext highlighter-rouge">SpeciesTree_Gene_Duplications_0.5_Support.txt</code> provides a summation of the above duplications over the branches of the species tree.</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Orthogroup_Sequences</code></p>
<ul>
  <li>A FASTA file for each orthogroup giving the amino acid sequences for each gene in the orthogroup.</li>
</ul>

<h2 id="latest-additions">Latest additions</h2>
<p>The current version of OrthoFinder has several major changes comapred to OrthoFinder version 2 (Emms &amp; Kelly 2019)</p>

<p><strong>New workflow for scalability</strong></p>

<p>The <code class="language-plaintext highlighter-rouge">--core --assign</code> workflow uses <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02652-8">SHOOT</a> to create profiles for previously computed orthogroups, and adds new genes to these orthogroups without requiring a costly all-versus-all sequence search. Genes that cannot be assigned using the SHOOT approach are analysed using a standard OrthoFinder workflow.</p>

<p><strong>Phylogenetic Hierarchical Orthogroups</strong></p>

<p>OrthoFinder has now extended its phylogenetic approach to orthogroups, allowing orthogroups to be defined for each node within the species tree. This significantly increases the accuracy of orthogroups, and enables users to perform orthogroup analyses for any clade of species in the species tree.</p>

<p align="center">
  <img src="assets/images/hog.png" alt="HOGs" width="400" />
</p>

<h2 id="citation">Citation</h2>

<p>The manuscript “OrthoFinder: scalable phylogenetic orthology inference for comparative genomics” is available as a preprint on biorxiv.</p>

<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0721-2]">Emms &amp; Kelly (2015)</a> introduced the orthogroup inference method.</p>

<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1832-y">Emms &amp; Kelly (2019)</a> introduced the phylogenetic inference of orthologs, including rooted gene and species trees, and gene duplication events.</p>

<p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5850722/">Emms &amp; Kelly (2017)</a> introduced the STRIDE method to root an unrooted species tree.</p>

<p><a href="https://www.biorxiv.org/content/10.1101/267914v1">Emms &amp; Kelly (2017)</a> introduced the STAG method of species tree inference.</p>

<h2 id="meet-the-team">Meet the team</h2>

<p>OrthoFinder was developed by David Emms &amp; Steve Kelly</p>

<p>Current members of the OrthoFinder team:</p>

<p>Yi Liu, Jonathan Holmes, Laurie Belcher</p>

<h2 id="how-to-cite">How to Cite?</h2>

<p>To cite RECUR, please cite both the following:</p>

<p>[1] <em>Elizabeth HJ Robbins, Yi Liu, Steven Kelly. 2025</em>. <strong>RECUR: Identifying recurrent amino acid substitutions from multiple sequence alignments</strong> bioRxiv. <a href="https://doi.org/10.1101/2025.04.29.651261"><img src="https://raw.githubusercontent.com/OrthoFinder/RECUR/main/docs/images/doi-badge.svg" alt="DOI:10.1101/2025.04.29.651261" /></a></p>

<p>[2] <em>Minh BQ, Schmidt HA, Chernomor O, Schrempf D, Woodhams MD, von Haeseler A, Lanfear R.</em> 2020. <strong>IQ-TREE 2: New Models and Efficient Methods for Phylogenetic Inference in the Genomic Era.</strong> Molecular Biology and Evolution 37:1530-1534. <a href="https://doi.org/10.1093/molbev/msaa015"><img src="https://raw.githubusercontent.com/OrthoFinder/RECUR/main/docs/images/doi-badge.svg" alt="DOI:10.1093/molbev/msaa015" /></a></p>

<p style="font-size: 18px;">
  If you want to cite an exemplar of RECUR’s usage, please cite:
</p>

<p>[3] <em>Robbins EHJ, Kelly S.</em> 2024. <strong>Widespread adaptive evolution in angiosperm photosystems provides insight into the evolution of photosystem II repair.</strong> The Plant Cell 37. <a href="https://doi.org/10.1093/plcell/koae281"><img src="https://raw.githubusercontent.com/OrthoFinder/RECUR/main/docs/images/doi-badge.svg" alt=" DOI: 10.1093/plcell/koae281" /></a></p>

 -->
  <div class="post-content">
    <h2 id="home">Home</h2>

<p align="center">
  <img src="assets/images/workflow.jpg" alt="workflow" width="1000" />
</p>
<p>OrthoFinder identifies orthogroups, infers gene trees for all orthogroups, and analyzes the gene trees to identify the rooted species tree. The method subsequently identifies all gene duplication events in the complete set of gene trees, and analyses them at both gene tree and species tree level. OrthoFinder further analyzes all of this phylogenetic information to identify the complete set of orthologs between all species, and provide extensive comparative genomics statistics.</p>

<h2 id="table-of-contents">Table of contents</h2>
<ul>
  <li><a href="#Installation">Installation</a></li>
  <li><a href="#Simple-Usage">Simple Usage</a></li>
  <li><a href="#Advanced-Usage-Scaling-to-Thousands-of-Species">Advanced Usage - Scaling to Thousands of Species</a></li>
  <li><a href="#Command-line-Options">Command line Options</a></li>
  <li><a href="#Output-files">Output files</a></li>
  <li><a href="#Latest-additions">Latest additions</a></li>
  <li><a href="#Citation">Citation</a></li>
</ul>

<p>A single PDF with all documentation and tutorial is available <a href="https://github.com/OrthoFinder/OrthoFinder/blob/update_readme/assets/docs/of3_v2.pdf">here</a></p>

<h2 id="installation">Installation</h2>

<h3 id="install-in-conda-recommended">Install in conda (recommended)</h3>

<p>The easiest way to install OrthoFinder is using <a href="https://www.machinelearningplus.com/deployment/conda-create-environment-and-everything-you-need-to-know-to-manage-conda-virtual-environment/">conda</a>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>conda create <span class="nt">-n</span> of3_env <span class="nv">python</span><span class="o">=</span>3.10
conda activate of3_env
conda <span class="nb">install </span>orthofinder
</code></pre></div></div>

<p>Alternatively, you could install via github, or download the source code and install locally.</p>

<h3 id="install-via-github">Install via github</h3>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>python3 <span class="nt">-m</span> venv of3_env 
<span class="nb">.</span> of3_env/bin/activate
pip <span class="nb">install </span>git+https://github.com/OrthoFinder/OrthoFinder.git
</code></pre></div></div>

<h3 id="install-locally-from-source-code">Install locally from source code</h3>

<p>The following commands provide three ways to download the source code of OrthoFinder locally into a directory named <code class="language-plaintext highlighter-rouge">OrthoFinder</code>.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Download via git </span>
git clone https://github.com/OrthoFinder/OrthoFinder.git
<span class="c"># or download the zipfile and unzip it into OrthoFinder</span>
<span class="nb">mkdir </span>OrthoFinder <span class="o">&amp;&amp;</span> wget <span class="nt">-qO-</span> https://github.com/OrthoFinder/OrthoFinder/archive/refs/heads/master.zip | funzip | <span class="nb">tar</span> <span class="nt">-x</span> <span class="nt">--strip-components</span><span class="o">=</span>1 <span class="nt">-C</span> OrthoFinder
<span class="c"># or download the tar.gz and unzip it into OrthoFinder</span>
<span class="nb">mkdir </span>OrthoFinder <span class="o">&amp;&amp;</span> wget <span class="nt">-qO-</span> https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.0.1/orthofinder-3.0.1.tar.gz | <span class="nb">tar</span> <span class="nt">-xz</span> <span class="nt">--strip-components</span><span class="o">=</span>1 <span class="nt">-C</span> OrthoFinder
</code></pre></div></div>

<p>Next, you can run the following commands to install OrthoFinder inside the of3_env virtural environment.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cd </span>OrthoFinder
python3 <span class="nt">-m</span> venv of3_env <span class="o">&amp;&amp;</span> <span class="nb">.</span> of3_env/bin/activate
pip <span class="nb">install</span> <span class="nb">.</span>
</code></pre></div></div>

<p>Whether you’ve installed OrthoFinder directly from GitHub or downloaded and set it up locally, the OrthoFinder package will only be available within the <code class="language-plaintext highlighter-rouge">of3_env</code> virtual environment. This avoids potential conflicts with Python dependencies.</p>

<p>To deactivate the virtual environment when you are finished, run:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>deactivate
</code></pre></div></div>

<h3 id="test-your-installation">Test your installation</h3>

<p>Once you have installed OrthoFinder, you can print the help information and version, and test it on the <a href="https://github.com/OrthoFinder/OrthoFinder/tree/master/ExampleData">example data</a>.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>orthofinder <span class="nt">--help</span> <span class="c"># Print out help informatioin</span>
orthofinder <span class="nt">--version</span> <span class="c"># Check the version</span>
orthofinder <span class="nt">-f</span> ExampleData <span class="c"># Test OrthoFinder on an example dataset</span>
</code></pre></div></div>

<h3 id="uninstalling">Uninstalling</h3>
<p>To uninstall on conda:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>conda deactivate
conda remove <span class="nt">-n</span> of3_env <span class="nt">--all</span>
</code></pre></div></div>
<p>To remove the virtual environment where OrthoFinder is installed:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>deactivate
<span class="nb">cd</span> ..
<span class="nb">rm</span> <span class="nt">-rf</span> OrthoFinder
</code></pre></div></div>

<h2 id="simple-usage">Simple Usage</h2>

<p>Run OrthoFinder on FASTA format proteomes in <code class="language-plaintext highlighter-rouge">&lt;dir&gt;</code></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">orthofinder</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span> <span class="o">-</span><span class="n">f</span> <span class="o">&lt;</span><span class="nb">dir</span><span class="o">&gt;</span>
</code></pre></div></div>

<p>OrthoFinder requires one FASTA file for each species. Each file should contain the complete set of protein sequences from that species’ genome,  with a single representative sequence for each gene.</p>

<p>If your files have multiple transcript variants for each gene, then we provide a script <code class="language-plaintext highlighter-rouge">primary_transcripts.py</code> to extract the longest variant per gene. This script should be run on your files prior to running OrthoFinder;</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for </span>f <span class="k">in</span> <span class="k">*</span>fa <span class="p">;</span> <span class="k">do </span>python primary_transcript.py <span class="nv">$f</span> <span class="p">;</span> <span class="k">done</span>
</code></pre></div></div>

<h2 id="advanced-usage---scaling-to-thousands-of-species">Advanced Usage - Scaling to Thousands of Species</h2>

<p>If you are analysing &gt;100 species, we reccommend that you use the scalable implementation.</p>

<p>Add the files for 64 species into one directory <code class="language-plaintext highlighter-rouge">&lt;core&gt;</code>
Add the remaining files into another directory <code class="language-plaintext highlighter-rouge">&lt;additional&gt;</code></p>

<p>First, run OrthoFinder on the subset of 64 species</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">orthofinder</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span> <span class="o">-</span><span class="n">f</span> <span class="o">&lt;</span><span class="n">core</span><span class="o">&gt;</span>
</code></pre></div></div>

<p>Then, add the additional species to the results of the core run</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">orthofinder</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span> <span class="o">--</span><span class="n">assign</span> <span class="o">&lt;</span><span class="n">additional</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">core</span> <span class="o">&lt;</span><span class="n">Results_Dir</span><span class="o">&gt;</span>
</code></pre></div></div>

<p>To choose which 64 species to include in the core, aim to capture a broad range of the evolutionary diversity of your species.</p>

<p align="center">
  <img src="assets/images/core_assign_figure.jpg" alt="core_assign_figure" width="800" />
</p>

<p><strong>Note that this alternative way of running OrthoFinder requires that the core species are run using the multiple sequence alignment option. You cannot add additional species to OrthoFinder results that were run with the <code class="language-plaintext highlighter-rouge">-M dendroblast</code> option, which was the default for OrthoFinder2</strong></p>

<h2 id="command-line-options">Command-line options</h2>

<p>Command-line options for OrthoFinder</p>

<p><strong>Adding additional species</strong>
| Parameter | Description                               |
|———–|——————————————-|
| <code class="language-plaintext highlighter-rouge">--assign &lt;dir1&gt; --core &lt;dir2&gt;</code>      | Assign species from <code class="language-plaintext highlighter-rouge">&lt;dir1&gt;</code> to existing orthogroups in <code class="language-plaintext highlighter-rouge">&lt;dir2&gt;</code>.                   |</p>

<p><strong>Method choices</strong>
| Parameter | Description                               | Default   | Options                                                                                     |
|———–|——————————————-|———–|———————————————————————————————|
| <code class="language-plaintext highlighter-rouge">-M</code>      | Method for gene tree inference.           | <code class="language-plaintext highlighter-rouge">msa</code>     | <code class="language-plaintext highlighter-rouge">dendroblast</code>, <code class="language-plaintext highlighter-rouge">msa</code>                                                                        |
| <code class="language-plaintext highlighter-rouge">-S</code>      | Sequence search program                   | <code class="language-plaintext highlighter-rouge">diamond</code> | <code class="language-plaintext highlighter-rouge">blast</code>, <code class="language-plaintext highlighter-rouge">diamond</code>, <code class="language-plaintext highlighter-rouge">diamond_ultra_sens</code>, <code class="language-plaintext highlighter-rouge">blastp</code>, <code class="language-plaintext highlighter-rouge">mmseqs</code>, <code class="language-plaintext highlighter-rouge">blastn</code> |
| <code class="language-plaintext highlighter-rouge">-A</code>      | MSA program, requires <code class="language-plaintext highlighter-rouge">-M msa</code>            | <code class="language-plaintext highlighter-rouge">famsa</code>   | <code class="language-plaintext highlighter-rouge">famsa</code>, <code class="language-plaintext highlighter-rouge">mafft</code>, <code class="language-plaintext highlighter-rouge">muscle</code>,                                                          |
| <code class="language-plaintext highlighter-rouge">-T</code>      | Tree inference method, requires <code class="language-plaintext highlighter-rouge">-M msa</code>  | <code class="language-plaintext highlighter-rouge">fasttree</code>| <code class="language-plaintext highlighter-rouge">fasttree</code>, <code class="language-plaintext highlighter-rouge">fasttree_fastest</code>, <code class="language-plaintext highlighter-rouge">raxml</code>, <code class="language-plaintext highlighter-rouge">iqtree</code>                               |
| <code class="language-plaintext highlighter-rouge">-I</code>      | MCL inflation parameter                   | <code class="language-plaintext highlighter-rouge">1.2</code>     | <code class="language-plaintext highlighter-rouge">1-10</code>                                                                                         |</p>

<p><strong>Input options</strong>
| Parameter | Description                               |
|———–|——————————————-|
| <code class="language-plaintext highlighter-rouge">-d</code>      | Input is DNA sequences.                   |
| <code class="language-plaintext highlighter-rouge">-s</code>      | User-specified rooted species tree.        |</p>

<p><strong>Output options</strong>
| Parameter   | Description                                                                 |
|———–  |—————————————————————————–|
| <code class="language-plaintext highlighter-rouge">-X</code>      | Don’t add species names to sequence IDs.                                    |
| <code class="language-plaintext highlighter-rouge">-n &lt;txt&gt;</code>      | Name to append to the results directory.                                    |
| <code class="language-plaintext highlighter-rouge">-o &lt;txt&gt;</code>      | Specify a non-default results directory.                                    |</p>

<p><strong>Parallel processing options</strong>
| Parameter | Description                                 | Default |
|———–|———————————————|———|
| <code class="language-plaintext highlighter-rouge">-t</code>      | Number of parallel sequence search threads. | <code class="language-plaintext highlighter-rouge">All available</code>    |
| <code class="language-plaintext highlighter-rouge">-a</code>      | Number of parallel analysis threads.        | <code class="language-plaintext highlighter-rouge">16 or t/8 (whichever lower)</code>     |</p>

<p><strong>Workflow stopping options</strong>
| Parameter | Description                                                                 |
|———–|—————————————————————————–|
| <code class="language-plaintext highlighter-rouge">-op</code>     | Stop after preparing input files for BLAST.                                 |</p>

<p><strong>Workflow restart options</strong>
| Parameter  | Description                                                  |
|————|————————————————————–|
| <code class="language-plaintext highlighter-rouge">-b &lt;dir&gt;</code> | Start OrthoFinder from pre-computed BLAST results in <code class="language-plaintext highlighter-rouge">&lt;dir&gt;</code>. |</p>

<p><strong>Other options</strong>
| Parameter        | Description                                                               |
|——————|—————————————————————————|
| <code class="language-plaintext highlighter-rouge">-1</code>             | Only perform one-way sequence search.                                     |
| <code class="language-plaintext highlighter-rouge">-z</code>             | Don’t trim MSAs (columns &gt;= 90% gap, min. alignment length 500).          |
| <code class="language-plaintext highlighter-rouge">-y</code>             | Split paralogous clades below the root of a HOG into separate HOGs.        |
| <code class="language-plaintext highlighter-rouge">-h</code>             | Print this help text.                                                     |
| <code class="language-plaintext highlighter-rouge">-v</code>             | Print version.                                                     |</p>

<h2 id="output-files">Output files</h2>

<p>A standard OrthoFinder run produces a set of files describing the orthogroups, orthologs, gene trees, resolve gene trees, the rooted species tree, gene duplication events, and comparative genomic statistics for the set of species being analysed. These files are located in an intuitive directory structure.</p>

<p>Full details on the output files and directories can be found in the PDF <a href="link">here</a>. The directories that are useful for most users are;</p>

<p><code class="language-plaintext highlighter-rouge">/Orthogroups</code></p>
<ul>
  <li>Orthogroups.tsv is the main orthogroup file. Each row contains the genes belonging to a single orthogroup. The genes from each orthogroup are organized into columns, one per species.</li>
  <li>Orthogroups.txt is a text file with each line showing the genes in a single orthogroup. It differs from Orthogroups.tsv in that it doesn’t show the species which each gene belongs to.</li>
  <li>Orthogroups.GeneCount.tsv is a tab separated text file that contains counts of the number of genes for each species in each orthogroup.</li>
  <li>Orthogroups_SingleCopyOrthologues.txt is a list of orthogroups that contain exactly one gene per species</li>
  <li>Orthogrouops_UnassignedGenes.tsv is a tab separated text file that contains all of the genes that were not assigned to any orthogroup.</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Phylogenetic_Hierarchical_Orthogroups</code></p>
<ul>
  <li>Each file is a phylogenetic hierarchical orthogroup (HOG) for a different node of the species tree</li>
  <li>Each row of a file contain the genes belonging to a single orthogroup</li>
  <li>Each species is represented by a single column</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Orthologues</code></p>
<ul>
  <li>Each species has a sub-directory that in turn contains a file for each pairwise species comparison, listing the orthologs between that species pair.</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Comparative_Genomics_Statistics</code></p>
<ul>
  <li>Files containing summary statistics across all orthogroups, as well as comparisons between each pair of species</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Resolved_Gene_Trees</code></p>
<ul>
  <li>A rooted phylogenetic tree inferred for each orthogroup with 4 or more sequences and resolved using the OrthoFinder hybrid species-overlap/duplication-loss coalescent model.</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Species_Tree</code></p>
<ul>
  <li>SpeciesTree_rooted.txt = A species tree inferred using ASTRAL-Pro.</li>
  <li>SpeciesTree_rooted_node_labels.txt = The same tree, but with nodes labels instead of support values. This labelled version is useful for interpreting and analysing the results of the gene duplication analyses.</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Gene_Duplication_Events</code></p>
<ul>
  <li><code class="language-plaintext highlighter-rouge">Duplications.tsv</code> has a row for each gene duplication event, with information on orthogroup in which it occured, the species that contain the duplicated gene, the node in the species tree on which the gene duplication event occured, and the support score for the gene duplication event.</li>
  <li><code class="language-plaintext highlighter-rouge">SpeciesTree_Gene_Duplications_0.5_Support.txt</code> provides a summation of the above duplications over the branches of the species tree.</li>
</ul>

<p><code class="language-plaintext highlighter-rouge">/Orthogroup_Sequences</code></p>
<ul>
  <li>A FASTA file for each orthogroup giving the amino acid sequences for each gene in the orthogroup.</li>
</ul>

<h2 id="latest-additions">Latest additions</h2>
<p>The current version of OrthoFinder has several major changes comapred to OrthoFinder version 2 (Emms &amp; Kelly 2019)</p>

<p><strong>New workflow for scalability</strong></p>

<p>The <code class="language-plaintext highlighter-rouge">--core --assign</code> workflow uses <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02652-8">SHOOT</a> to create profiles for previously computed orthogroups, and adds new genes to these orthogroups without requiring a costly all-versus-all sequence search. Genes that cannot be assigned using the SHOOT approach are analysed using a standard OrthoFinder workflow.</p>

<p><strong>Phylogenetic Hierarchical Orthogroups</strong></p>

<p>OrthoFinder has now extended its phylogenetic approach to orthogroups, allowing orthogroups to be defined for each node within the species tree. This significantly increases the accuracy of orthogroups, and enables users to perform orthogroup analyses for any clade of species in the species tree.</p>

<p align="center">
  <img src="assets/images/hog.png" alt="HOGs" width="400" />
</p>

<h2 id="citation">Citation</h2>

<p>The manuscript “OrthoFinder: scalable phylogenetic orthology inference for comparative genomics” is available as a preprint on biorxiv.</p>

<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0721-2]">Emms &amp; Kelly (2015)</a> introduced the orthogroup inference method.</p>

<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1832-y">Emms &amp; Kelly (2019)</a> introduced the phylogenetic inference of orthologs, including rooted gene and species trees, and gene duplication events.</p>

<p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5850722/">Emms &amp; Kelly (2017)</a> introduced the STRIDE method to root an unrooted species tree.</p>

<p><a href="https://www.biorxiv.org/content/10.1101/267914v1">Emms &amp; Kelly (2017)</a> introduced the STAG method of species tree inference.</p>

<h2 id="meet-the-team">Meet the team</h2>

<p>OrthoFinder was developed by David Emms &amp; Steve Kelly</p>

<p>Current members of the OrthoFinder team:</p>

<p>Yi Liu, Jonathan Holmes, Laurie Belcher</p>

<h2 id="how-to-cite">How to Cite?</h2>

<p>To cite RECUR, please cite both the following:</p>

<p>[1] <em>Elizabeth HJ Robbins, Yi Liu, Steven Kelly. 2025</em>. <strong>RECUR: Identifying recurrent amino acid substitutions from multiple sequence alignments</strong> bioRxiv. <a href="https://doi.org/10.1101/2025.04.29.651261"><img src="https://raw.githubusercontent.com/OrthoFinder/RECUR/main/docs/images/doi-badge.svg" alt="DOI:10.1101/2025.04.29.651261" /></a></p>

<p>[2] <em>Minh BQ, Schmidt HA, Chernomor O, Schrempf D, Woodhams MD, von Haeseler A, Lanfear R.</em> 2020. <strong>IQ-TREE 2: New Models and Efficient Methods for Phylogenetic Inference in the Genomic Era.</strong> Molecular Biology and Evolution 37:1530-1534. <a href="https://doi.org/10.1093/molbev/msaa015"><img src="https://raw.githubusercontent.com/OrthoFinder/RECUR/main/docs/images/doi-badge.svg" alt="DOI:10.1093/molbev/msaa015" /></a></p>

<p style="font-size: 18px;">
  If you want to cite an exemplar of RECUR’s usage, please cite:
</p>

<p>[3] <em>Robbins EHJ, Kelly S.</em> 2024. <strong>Widespread adaptive evolution in angiosperm photosystems provides insight into the evolution of photosystem II repair.</strong> The Plant Cell 37. <a href="https://doi.org/10.1093/plcell/koae281"><img src="https://raw.githubusercontent.com/OrthoFinder/RECUR/main/docs/images/doi-badge.svg" alt=" DOI: 10.1093/plcell/koae281" /></a></p>


  </div>

  


  <!---->
  <!-- <ul class="post-list"><li>
        <span class="post-meta">May 7, 2025</span>
        <h3>
          <a class="post-link" href="/OrthoFinder/release/07/05/2025/OrthoFinder-v3.1.0.html">
            OrthoFinder-v3.1.0
          </a>
        </h3><div class="download-item">
  📦 
    <a href="https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/orthofinder-linux-intel-3.1.0.tar.gz" class="btn">
        orthofinder-linux-intel-3.1.0.tar.gz
    </a>
    <span class="download-count" data-asset="orthofinder-linux-intel-3.1.0.tar.gz">
    <i class="fa fa-download" aria-hidden="true"></i>
    </span>
</div>
<div class="download-item">
  📦 
    <a href="https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/orthofinder-linux-arm-3.1.0.tar.gz" class="btn">
        orthofinder-linux-arm-3.1.0.tar.gz
    </a>
    <span class="download-count" data-asset="orthofinder-linux-arm-3.1.0.tar.gz">
    <i class="fa fa-download" aria-hidden="true"></i>
    </span>
</div>
<div class="download-item">
  📦 
    <a href="https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/orthofinder-macos-arm-3.1.0.tar.gz" class="btn">
        orthofinder-macos-arm-3.1.0.tar.gz
    </a>
    <span class="download-count" data-asset="orthofinder-macos-arm-3.1.0.tar.gz">
    <i class="fa fa-download" aria-hidden="true"></i>
    </span>
</div>
<div class="download-item">
  📦 
    <a href="https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/orthofinder-macos-intel-3.1.0.tar.gz" class="btn">
        orthofinder-macos-intel-3.1.0.tar.gz
    </a>
    <span class="download-count" data-asset="orthofinder-macos-intel-3.1.0.tar.gz">
    <i class="fa fa-download" aria-hidden="true"></i>
    </span>
</div>
<div class="download-item">
  📦 
    <a href="https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/orthofinder-macos-universal-3.1.0.tar.gz" class="btn">
        orthofinder-macos-universal-3.1.0.tar.gz
    </a>
    <span class="download-count" data-asset="orthofinder-macos-universal-3.1.0.tar.gz">
    <i class="fa fa-download" aria-hidden="true"></i>
    </span>
</div>

</li></ul> -->

  </div>
        
      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/OrthoFinder/"></data>

  <div class="wrapper">

    <div class="footer-col-wrapper">
      <div class="footer-col">

        <!-- <p class="feed-subscribe">
          <a href="http://localhost:4000/OrthoFinder/feed.xml">
            <svg class="svg-icon orange">
              <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#rss"></use>
            </svg><span>Subscribe</span>
          </a>
        </p> -->

        <!--<ul class="social-media-list"><li>
  <a rel="me" href="" target="_blank" title="GitHub">
    <svg class="svg-icon grey">
      <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
    </svg>
  </a>
</li>


</ul>
-->
         
        <div class="social-icons">
          
          <a href="https://github.com/OrthoFinder/OrthoFinder" target="_blank" title="GitHub">
            <svg class="svg-icon grey">
              <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
            </svg>
            <span>GitHub</span>
          </a>
          
          <!--  -->
        </div>


          <!--
          <ul class="contact-list">

            Link to the team section instead of listing every author
            <li>
              <a href="/OrthoFinder/about/#team">
                Meet&nbsp;the&nbsp;Team
              </a>
            </li>

          </ul> -->

          <ul class="contact-list">
            <li><a href="/OrthoFinder/about/#team">Meet the Team</a></li>
          </ul>


      </div>
      
      <div class="footer-col">
        <!-- <p>This is a software developed by the [Steven Kelly Lab](http://www.stevekellylab.com/).</p> -->
        <p><p>This is a software developed by the <a href="http://www.stevekellylab.com/">Steven Kelly Lab</a>.</p>
</p>
      </div>
    </div>

    <!-- <div class="social-links"><ul class="social-media-list"><li>
  <a rel="me" href="" target="_blank" title="GitHub">
    <svg class="svg-icon grey">
      <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
    </svg>
  </a>
</li>


</ul>
</div> -->

  </div>


<script>
  document.addEventListener('DOMContentLoaded', function () {
    // 1) Find all <pre><code> blocks
    const codeBlocks = document.querySelectorAll('pre > code');

    codeBlocks.forEach((codeBlock, index) => {
      // The parent <pre> element
      const pre = codeBlock.parentNode;

      // 2) Create a container DIV and wrap <pre> inside it
      const container = document.createElement('div');
      container.className = 'code-container';
      pre.parentNode.insertBefore(container, pre);
      container.appendChild(pre);

      // 3) Create the copy button
      const button = document.createElement('button');
      button.className = 'copy-button';
      button.type = 'button';

      // 4) Insert an inline SVG icon + "Copy" text
      //    We'll build the SVG as a string, then inject it into button.innerHTML
      const svgIcon = `
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="16"
          height="16"
          fill="currentColor"
          class="bi bi-files"
          viewBox="0 0 16 16"
        >
            <path d="M13 0H5a1 1 0 0 0-1 1v1H3a2 2 0 0 0-2
                  2v9.5A1.5 1.5 0 0 0 2.5 15H9v-1H2.5a.5.5
                  0 0 1-.5-.5V4a1 1 0 0 1 1-1h1v7.5A1.5
                  1.5 0 0 0 5.5 12H11v-1H5.5a.5.5 0 0
                  1-.5-.5V1a.5.5 0 0 1 .5-.5H5v3.5A1.5
                  1.5 0 0 0 6.5 5H13v6h1V1a1 1 0 0 0-1-1z"/>
        </svg>
    `;
      

      
      button.innerHTML = svgIcon + 'Copy';

      // 5) Assign an ID to <code> if it doesn't have one
      if (!codeBlock.id) {
        codeBlock.id = `codeblock-${index}`;
      }

      // 6) Link button to the code block by ID
      button.setAttribute('data-clipboard-target', '#' + codeBlock.id);

      // 7) Insert the button into the container
      container.insertBefore(button, pre);
    });

    // 8) Initialize Clipboard.js for all .copy-button elements
    const clipboard = new ClipboardJS('.copy-button');
    clipboard.on('success', function (e) {
      e.clearSelection();
      // Briefly change button text to indicate success
      e.trigger.innerText = 'Copied!';
      setTimeout(() => {
        e.trigger.innerHTML = svgIcon + 'Copy'; // restore icon + text
      }, 1500);
    });
    clipboard.on('error', function (e) {
      console.error('Failed to copy', e);
    });
  });
</script>


  <!-- <script>
    document.addEventListener('DOMContentLoaded', function () {
      // 1. Find all <pre><code> blocks
      const codeBlocks = document.querySelectorAll('pre > code');

      codeBlocks.forEach((codeBlock) => {
        const pre = codeBlock.parentNode;

        // 2. Create a container div to hold the code + button
        const container = document.createElement('div');
        container.className = 'code-container';

        // Insert the container before <pre>, then move <pre> inside it
        pre.parentNode.insertBefore(container, pre);
        container.appendChild(pre);

        // 3. Create a "Copy" button
        const button = document.createElement('button');
        button.className = 'copy-button';
        button.type = 'button';
        button.innerText = 'Copy';

        // If the code block doesn’t have an ID, assign one so we can target it
        let codeId = codeBlock.id;
        if (!codeId) {
          codeId = 'codeblock-' + Math.floor(Math.random() * 1000000);
          codeBlock.setAttribute('id', codeId);
        }

        // 4. Point the button at the code block
        button.setAttribute('data-clipboard-target', '#' + codeId);

        // 5. Insert button into the container
        container.insertBefore(button, pre);
      });

      // 6. Initialize Clipboard.js
      const clipboard = new ClipboardJS('.copy-button');
      clipboard.on('success', function (e) {
        e.clearSelection();
        // Change button text to confirm copy
        e.trigger.textContent = 'Copied!';
        setTimeout(() => {
          e.trigger.textContent = 'Copy';
        }, 2000);
      });
      clipboard.on('error', function (e) {
        console.error('Failed to copy', e);
      });
    });
  </script> -->


</footer></body>

</html>

