<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Guid to Results | OrthoFinder</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="Guid to Results" />
<meta name="author" content="David Emms, Jonathan Holmes, Laurence Belcher, Yi Liu" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="A complete guide to OrthoFinder results files Definitions for some terms that are used in these files: Species-specific orthogroup: An orthogroups that consist entirely of genes from one species. G50: The number of genes in the orthogroup such that 50% of genes are in orthogroups of that size or larger. O50: The smallest number of orthogroups such that 50% of genes are in orthogroups of that size or larger. Single-copy orthogroup: An orthogroup with exactly one gene (and no more) from each species. Unassigned gene: A gene that has not been put into an orthogroup with any other genes. Comparative_Genomics_Statistics Duplications_per_Orthogroup.tsv is a tab separated text file that gives the number of duplications identified in each orthogroup. Duplications_per_Species_Tree_Node.tsv is a tab separated text file that gives the number of duplications identified as occurring along each branch of the species tree. Orthogroups_SpeciesOverlaps.tsv is a tab separated text file that contains the number of orthogroups shared between each species-pair as a square matrix. OrthologuesStats_*.tsv files are tab separated text files containing matrices giving the numbers of orthologues in one-to-one, one-to-many and many-to-many relationships between each pair of species. OrthologuesStats_one-to-one.tsv is the number of one-to-one orthologues between each species pair. OrthologuesStats_many-to-many.tsv contains the number of orthologues in a many-to-many relationship for each species pair (due to gene duplication events in both lineages post-speciation). Entry (i,j) is the number of genes in species i that are in a many-to-many orthology relationship with genes in species j. OrthologuesStats_one-to-many.tsv: entry (i,j) gives the number of genes in species i that are in a one-to-many orthology relationship with genes from species j. OrthologuesStats_many-to-one.tsv: entry (i,j) gives the number of genes in species i that are in a many-to-one orthology relationship with a gene from species j. OrthologuesStats_Total.tsv contains the totals for each species pair of orthologues of whatever multiplicity. Entry (i,j) is the total number of genes in species i that have orthologues in species j. Statistics_Overall.tsv is a tab separated text file that contains general statistics about orthogroup sizes and proportion of genes assigned to orthogroups. Statistics_PerSpecies.tsv is a tab separated text file that contains the same information as the Statistics_Overall.csv file but for each individual species. Gene_Duplication_Events Duplications.tsv is a tab separated text file that lists all the gene duplication events identified by examining each node of each orthogroup gene tree. The columns are; a. “Orthogroup” b. “Species Tree node” (see Species_Tree/SpeciesTree_rooted_node_labels.txt) c. “Gene tree node” (see corresponding orthogroup tree in Resolved_Gene_Trees/) d. “Support” (proportion of expected species for which both copies of the duplicated gene are present) e. “Type” i. “Terminal”: duplication on a terminal branch of the species tree ii. “Non-Terminal”: duplication on an internal branch of the species tree &amp; therefore shared by more than one species iii. “Non-Terminal: STRIDE”: Non-Terminal duplication that also passes the very stringent STRIDE checks for what the topology of the gene tree should be post-duplication) f. “Genes 1” (the list of genes descended from one of the copies of the duplicate gene) g. “Genes 2” (the list of genes descended from the other copy of the duplicate gene. SpeciesTree_Gene_Duplications_0.5_Support.txt provides a summation of the above duplications over the branches of the species tree. The numbers after each node or species name are the number of gene duplication events with at least 50% support that occurred on the branch leading to the node/species. The branch lengths are as given in Species_Tree/SpeciesTree_rooted.txt. It is a text file, in newick format. Orthogroup Sequences A FASTA file for each orthogroup giving the amino acid sequences for each gene in the orthogroup. Orthogroups Directory Orthogroups.tsv is a tab separated text file that contains the genes in each orthogroup, with columns for each species. Orthogroups.txt is a text file that contains the genes in each orthogroup (one orthogroup per line). Orthogroups_UnassignedGenes.tsv is a tab separated text file that contains all of the genes that were not assigned to any orthogroup. Orthogroups.GeneCount.tsv is a tab separated text file that contains counts of the number of genes for each species in each orthogroup. Orthogroups_SingleCopyOrthologues.txt is a list of orthogroups that contain exactly one gene per species i.e. they contain one-to-one orthologues. Orthologues Directory One sub-directory for each species that in turn contains a file for each pairwise species comparison, listing the orthologs between that species pair. Orthologues can be one-to-one, one-to-many or many-to-many depending on the gene duplication events since the orthologs diverged. Each row in a file contains the gene(s) in one species that are orthologues of the gene(s) in the other species and each row is cross-referenced to the orthogroup that contains those genes. Phylogenetic Hierarchical Orthogroups Directory OrthoFinder infers HOGs, orthogroups at each hierarchical level (i.e. at each node in the species tree) by analysing the rooted gene trees. N1.tsv is a tab separated text file. Each row contains the genes belonging to a single orthogroup. The genes from each orthogroup are organized into columns, one per species. Additional columns give the HOG (Hierarchical Orthogroup) ID and the node in the gene tree from which the HOG was determined N2.tsv, N3.tsv, …: Orthogroups inferred from the gene trees corresponding to the clades of species in the species tree N2, N3, etc. Phylogenetically_Misplaced_Genes Genes in “Phylogenetically_Misplaced_Genes/” are those that appear to be out of place in the gene tree, and would otherwise negatively affect orthology analysis if not identified. Putative_Xenologs Xenologs are sets of genes descended from a common ancestor, but where there has been horizontal transfer on the evolutionary path to the gene copies in extant species, rather than just speciation and duplication. OrthoFinder tries to identify xenologs, but we call them ‘putative’, since many arise from contamination during sequencing. Each species has a file in this folder, listing genes and their putative xenologs from all other species Resolved Gene Trees Directory A rooted phylogenetic tree inferred for each orthogroup with 4 or more sequences and resolved using the OrthoFinder hybrid species-overlap/duplication-loss coalescent model. Single Copy Orthologue Sequences The same files as the “Orthogroup Sequences” directory but restricted to only those orthogroups which contain exactly one gene per species. Species Tree Directory SpeciesTree_rooted.txt: A STAG species tree inferred from all orthogroups, containing STAG support values at internal nodes and rooted using STRIDE. SpeciesTree_rooted_node_labels.txt: The same tree as above but with the nodes given labels (instead of support values) to allow other results files to crossreference branches/nodes in the species tree (e.g. location of gene duplication events). WorkingDirectory This contains all the files necessary for orthofinder to run" />
<meta property="og:description" content="A complete guide to OrthoFinder results files Definitions for some terms that are used in these files: Species-specific orthogroup: An orthogroups that consist entirely of genes from one species. G50: The number of genes in the orthogroup such that 50% of genes are in orthogroups of that size or larger. O50: The smallest number of orthogroups such that 50% of genes are in orthogroups of that size or larger. Single-copy orthogroup: An orthogroup with exactly one gene (and no more) from each species. Unassigned gene: A gene that has not been put into an orthogroup with any other genes. Comparative_Genomics_Statistics Duplications_per_Orthogroup.tsv is a tab separated text file that gives the number of duplications identified in each orthogroup. Duplications_per_Species_Tree_Node.tsv is a tab separated text file that gives the number of duplications identified as occurring along each branch of the species tree. Orthogroups_SpeciesOverlaps.tsv is a tab separated text file that contains the number of orthogroups shared between each species-pair as a square matrix. OrthologuesStats_*.tsv files are tab separated text files containing matrices giving the numbers of orthologues in one-to-one, one-to-many and many-to-many relationships between each pair of species. OrthologuesStats_one-to-one.tsv is the number of one-to-one orthologues between each species pair. OrthologuesStats_many-to-many.tsv contains the number of orthologues in a many-to-many relationship for each species pair (due to gene duplication events in both lineages post-speciation). Entry (i,j) is the number of genes in species i that are in a many-to-many orthology relationship with genes in species j. OrthologuesStats_one-to-many.tsv: entry (i,j) gives the number of genes in species i that are in a one-to-many orthology relationship with genes from species j. OrthologuesStats_many-to-one.tsv: entry (i,j) gives the number of genes in species i that are in a many-to-one orthology relationship with a gene from species j. OrthologuesStats_Total.tsv contains the totals for each species pair of orthologues of whatever multiplicity. Entry (i,j) is the total number of genes in species i that have orthologues in species j. Statistics_Overall.tsv is a tab separated text file that contains general statistics about orthogroup sizes and proportion of genes assigned to orthogroups. Statistics_PerSpecies.tsv is a tab separated text file that contains the same information as the Statistics_Overall.csv file but for each individual species. Gene_Duplication_Events Duplications.tsv is a tab separated text file that lists all the gene duplication events identified by examining each node of each orthogroup gene tree. The columns are; a. “Orthogroup” b. “Species Tree node” (see Species_Tree/SpeciesTree_rooted_node_labels.txt) c. “Gene tree node” (see corresponding orthogroup tree in Resolved_Gene_Trees/) d. “Support” (proportion of expected species for which both copies of the duplicated gene are present) e. “Type” i. “Terminal”: duplication on a terminal branch of the species tree ii. “Non-Terminal”: duplication on an internal branch of the species tree &amp; therefore shared by more than one species iii. “Non-Terminal: STRIDE”: Non-Terminal duplication that also passes the very stringent STRIDE checks for what the topology of the gene tree should be post-duplication) f. “Genes 1” (the list of genes descended from one of the copies of the duplicate gene) g. “Genes 2” (the list of genes descended from the other copy of the duplicate gene. SpeciesTree_Gene_Duplications_0.5_Support.txt provides a summation of the above duplications over the branches of the species tree. The numbers after each node or species name are the number of gene duplication events with at least 50% support that occurred on the branch leading to the node/species. The branch lengths are as given in Species_Tree/SpeciesTree_rooted.txt. It is a text file, in newick format. Orthogroup Sequences A FASTA file for each orthogroup giving the amino acid sequences for each gene in the orthogroup. Orthogroups Directory Orthogroups.tsv is a tab separated text file that contains the genes in each orthogroup, with columns for each species. Orthogroups.txt is a text file that contains the genes in each orthogroup (one orthogroup per line). Orthogroups_UnassignedGenes.tsv is a tab separated text file that contains all of the genes that were not assigned to any orthogroup. Orthogroups.GeneCount.tsv is a tab separated text file that contains counts of the number of genes for each species in each orthogroup. Orthogroups_SingleCopyOrthologues.txt is a list of orthogroups that contain exactly one gene per species i.e. they contain one-to-one orthologues. Orthologues Directory One sub-directory for each species that in turn contains a file for each pairwise species comparison, listing the orthologs between that species pair. Orthologues can be one-to-one, one-to-many or many-to-many depending on the gene duplication events since the orthologs diverged. Each row in a file contains the gene(s) in one species that are orthologues of the gene(s) in the other species and each row is cross-referenced to the orthogroup that contains those genes. Phylogenetic Hierarchical Orthogroups Directory OrthoFinder infers HOGs, orthogroups at each hierarchical level (i.e. at each node in the species tree) by analysing the rooted gene trees. N1.tsv is a tab separated text file. Each row contains the genes belonging to a single orthogroup. The genes from each orthogroup are organized into columns, one per species. Additional columns give the HOG (Hierarchical Orthogroup) ID and the node in the gene tree from which the HOG was determined N2.tsv, N3.tsv, …: Orthogroups inferred from the gene trees corresponding to the clades of species in the species tree N2, N3, etc. Phylogenetically_Misplaced_Genes Genes in “Phylogenetically_Misplaced_Genes/” are those that appear to be out of place in the gene tree, and would otherwise negatively affect orthology analysis if not identified. Putative_Xenologs Xenologs are sets of genes descended from a common ancestor, but where there has been horizontal transfer on the evolutionary path to the gene copies in extant species, rather than just speciation and duplication. OrthoFinder tries to identify xenologs, but we call them ‘putative’, since many arise from contamination during sequencing. Each species has a file in this folder, listing genes and their putative xenologs from all other species Resolved Gene Trees Directory A rooted phylogenetic tree inferred for each orthogroup with 4 or more sequences and resolved using the OrthoFinder hybrid species-overlap/duplication-loss coalescent model. Single Copy Orthologue Sequences The same files as the “Orthogroup Sequences” directory but restricted to only those orthogroups which contain exactly one gene per species. Species Tree Directory SpeciesTree_rooted.txt: A STAG species tree inferred from all orthogroups, containing STAG support values at internal nodes and rooted using STRIDE. SpeciesTree_rooted_node_labels.txt: The same tree as above but with the nodes given labels (instead of support values) to allow other results files to crossreference branches/nodes in the species tree (e.g. location of gene duplication events). WorkingDirectory This contains all the files necessary for orthofinder to run" />
<link rel="canonical" href="http://localhost:4000/OrthoFinder/tutorials/guide-to-results/" />
<meta property="og:url" content="http://localhost:4000/OrthoFinder/tutorials/guide-to-results/" />
<meta property="og:site_name" content="OrthoFinder" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2025-05-01T10:05:16+01:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Guid to Results" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","author":{"@type":"Person","name":"David Emms, Jonathan Holmes, Laurence Belcher, Yi Liu"},"dateModified":"2025-05-01T10:05:16+01:00","datePublished":"2025-05-01T10:05:16+01:00","description":"A complete guide to OrthoFinder results files Definitions for some terms that are used in these files: Species-specific orthogroup: An orthogroups that consist entirely of genes from one species. G50: The number of genes in the orthogroup such that 50% of genes are in orthogroups of that size or larger. O50: The smallest number of orthogroups such that 50% of genes are in orthogroups of that size or larger. Single-copy orthogroup: An orthogroup with exactly one gene (and no more) from each species. Unassigned gene: A gene that has not been put into an orthogroup with any other genes. Comparative_Genomics_Statistics Duplications_per_Orthogroup.tsv is a tab separated text file that gives the number of duplications identified in each orthogroup. Duplications_per_Species_Tree_Node.tsv is a tab separated text file that gives the number of duplications identified as occurring along each branch of the species tree. Orthogroups_SpeciesOverlaps.tsv is a tab separated text file that contains the number of orthogroups shared between each species-pair as a square matrix. OrthologuesStats_*.tsv files are tab separated text files containing matrices giving the numbers of orthologues in one-to-one, one-to-many and many-to-many relationships between each pair of species. OrthologuesStats_one-to-one.tsv is the number of one-to-one orthologues between each species pair. OrthologuesStats_many-to-many.tsv contains the number of orthologues in a many-to-many relationship for each species pair (due to gene duplication events in both lineages post-speciation). Entry (i,j) is the number of genes in species i that are in a many-to-many orthology relationship with genes in species j. OrthologuesStats_one-to-many.tsv: entry (i,j) gives the number of genes in species i that are in a one-to-many orthology relationship with genes from species j. OrthologuesStats_many-to-one.tsv: entry (i,j) gives the number of genes in species i that are in a many-to-one orthology relationship with a gene from species j. OrthologuesStats_Total.tsv contains the totals for each species pair of orthologues of whatever multiplicity. Entry (i,j) is the total number of genes in species i that have orthologues in species j. Statistics_Overall.tsv is a tab separated text file that contains general statistics about orthogroup sizes and proportion of genes assigned to orthogroups. Statistics_PerSpecies.tsv is a tab separated text file that contains the same information as the Statistics_Overall.csv file but for each individual species. Gene_Duplication_Events Duplications.tsv is a tab separated text file that lists all the gene duplication events identified by examining each node of each orthogroup gene tree. The columns are; a. “Orthogroup” b. “Species Tree node” (see Species_Tree/SpeciesTree_rooted_node_labels.txt) c. “Gene tree node” (see corresponding orthogroup tree in Resolved_Gene_Trees/) d. “Support” (proportion of expected species for which both copies of the duplicated gene are present) e. “Type” i. “Terminal”: duplication on a terminal branch of the species tree ii. “Non-Terminal”: duplication on an internal branch of the species tree &amp; therefore shared by more than one species iii. “Non-Terminal: STRIDE”: Non-Terminal duplication that also passes the very stringent STRIDE checks for what the topology of the gene tree should be post-duplication) f. “Genes 1” (the list of genes descended from one of the copies of the duplicate gene) g. “Genes 2” (the list of genes descended from the other copy of the duplicate gene. SpeciesTree_Gene_Duplications_0.5_Support.txt provides a summation of the above duplications over the branches of the species tree. The numbers after each node or species name are the number of gene duplication events with at least 50% support that occurred on the branch leading to the node/species. The branch lengths are as given in Species_Tree/SpeciesTree_rooted.txt. It is a text file, in newick format. Orthogroup Sequences A FASTA file for each orthogroup giving the amino acid sequences for each gene in the orthogroup. Orthogroups Directory Orthogroups.tsv is a tab separated text file that contains the genes in each orthogroup, with columns for each species. Orthogroups.txt is a text file that contains the genes in each orthogroup (one orthogroup per line). Orthogroups_UnassignedGenes.tsv is a tab separated text file that contains all of the genes that were not assigned to any orthogroup. Orthogroups.GeneCount.tsv is a tab separated text file that contains counts of the number of genes for each species in each orthogroup. Orthogroups_SingleCopyOrthologues.txt is a list of orthogroups that contain exactly one gene per species i.e. they contain one-to-one orthologues. Orthologues Directory One sub-directory for each species that in turn contains a file for each pairwise species comparison, listing the orthologs between that species pair. Orthologues can be one-to-one, one-to-many or many-to-many depending on the gene duplication events since the orthologs diverged. Each row in a file contains the gene(s) in one species that are orthologues of the gene(s) in the other species and each row is cross-referenced to the orthogroup that contains those genes. Phylogenetic Hierarchical Orthogroups Directory OrthoFinder infers HOGs, orthogroups at each hierarchical level (i.e. at each node in the species tree) by analysing the rooted gene trees. N1.tsv is a tab separated text file. Each row contains the genes belonging to a single orthogroup. The genes from each orthogroup are organized into columns, one per species. Additional columns give the HOG (Hierarchical Orthogroup) ID and the node in the gene tree from which the HOG was determined N2.tsv, N3.tsv, …: Orthogroups inferred from the gene trees corresponding to the clades of species in the species tree N2, N3, etc. Phylogenetically_Misplaced_Genes Genes in “Phylogenetically_Misplaced_Genes/” are those that appear to be out of place in the gene tree, and would otherwise negatively affect orthology analysis if not identified. Putative_Xenologs Xenologs are sets of genes descended from a common ancestor, but where there has been horizontal transfer on the evolutionary path to the gene copies in extant species, rather than just speciation and duplication. OrthoFinder tries to identify xenologs, but we call them ‘putative’, since many arise from contamination during sequencing. Each species has a file in this folder, listing genes and their putative xenologs from all other species Resolved Gene Trees Directory A rooted phylogenetic tree inferred for each orthogroup with 4 or more sequences and resolved using the OrthoFinder hybrid species-overlap/duplication-loss coalescent model. Single Copy Orthologue Sequences The same files as the “Orthogroup Sequences” directory but restricted to only those orthogroups which contain exactly one gene per species. Species Tree Directory SpeciesTree_rooted.txt: A STAG species tree inferred from all orthogroups, containing STAG support values at internal nodes and rooted using STRIDE. SpeciesTree_rooted_node_labels.txt: The same tree as above but with the nodes given labels (instead of support values) to allow other results files to crossreference branches/nodes in the species tree (e.g. location of gene duplication events). WorkingDirectory This contains all the files necessary for orthofinder to run","headline":"Guid to Results","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/OrthoFinder/tutorials/guide-to-results/"},"url":"http://localhost:4000/OrthoFinder/tutorials/guide-to-results/"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/OrthoFinder/assets/css/style.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/OrthoFinder/feed.xml" title="OrthoFinder" /><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
</head>
<body><header class="site-header">

  <div class="wrapper"><!-- <a class="site-title" rel="author" href="/OrthoFinder/%20/">OrthoFinder</a> -->
    <a class="site-title" rel="author" href="/OrthoFinder/">OrthoFinder</a><nav class="site-nav">
      <input type="checkbox" id="nav-trigger" class="nav-trigger" />
      <label for="nav-trigger">
        <span class="menu-icon">
          <svg viewBox="0 0 18 15" width="18px" height="15px">
            <path
              d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z" />
          </svg>
        </span>
      </label>
      <div class="trigger">
        <!--<a class="page-link" href="/OrthoFinder/">Home</a><a class="page-link" href="/OrthoFinder/about/">About</a><a class="page-link" href="/OrthoFinder/download/">Download</a><a class="page-link" href="/OrthoFinder/installation/">Installation</a><a class="page-link" href="/OrthoFinder/tutorial/">Tutorial</a><a class="page-link" href="/OrthoFinder/faq/">FAQ</a>--><a class="page-link" href="/OrthoFinder/">Home</a><a class="page-link" href="/OrthoFinder/about/">About</a><a class="page-link" href="/OrthoFinder/download/">Download</a><a class="page-link" href="/OrthoFinder/installation/">Installation</a><a class="page-link" href="/OrthoFinder/tutorial/">Tutorial</a><a class="page-link" href="/OrthoFinder/faq/">FAQ</a>
        <a class="page-link" href="https://github.com/OrthoFinder/OrthoFinder" target="_blank" title="GitHub Profile">GitHub</a>
        
      </div>
    </nav></div>
  <script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js"></script>
</header><main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Guid to Results</h1>
    <p class="post-meta"><!-- <time class="dt-published" datetime="2025-05-01T10:05:16+01:00" itemprop="datePublished">
        May 1, 2025
      </time> --><!-- <i class="fa fa-refresh" aria-hidden="true"></i> -->
      <time class="dt-modified" datetime="2025-05-06T00:00:00+01:00" itemprop="dateModified">
        Updated May 6, 2025
      </time>
      <meta itemprop="datePublished" content="2025-05-01T10:05:16+01:00"></p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <h2 id="a-complete-guide-to-orthofinder-results-files">A complete guide to OrthoFinder results files</h2>

<p><em>Definitions for some terms that are used in these files:</em></p>

<ul>
  <li>Species-specific orthogroup: An orthogroups that consist entirely of genes from
one species.</li>
  <li>G50: The number of genes in the orthogroup such that 50% of genes are in
orthogroups of that size or larger.</li>
  <li>O50: The smallest number of orthogroups such that 50% of genes are in orthogroups
of that size or larger.</li>
  <li>Single-copy orthogroup: An orthogroup with exactly one gene (and no more) from
each species.</li>
  <li>Unassigned gene: A gene that has not been put into an orthogroup with any other
genes.</li>
</ul>

<h2 id="comparative_genomics_statistics">Comparative_Genomics_Statistics</h2>

<ul>
  <li><strong>Duplications_per_Orthogroup.tsv</strong> is a tab separated text file that gives the number
of duplications identified in each orthogroup.</li>
  <li><strong>Duplications_per_Species_Tree_Node.tsv</strong> is a tab separated text file that gives the
number of duplications identified as occurring along each branch of the species tree.</li>
  <li><strong>Orthogroups_SpeciesOverlaps.tsv</strong> is a tab separated text file that contains the number of orthogroups shared between each species-pair as a square matrix.</li>
  <li><strong>OrthologuesStats_*.tsv</strong> files are tab separated text files containing matrices giving
the numbers of orthologues in one-to-one, one-to-many and many-to-many
relationships between each pair of species.
    <ul>
      <li><strong>OrthologuesStats_one-to-one.tsv</strong> is the number of one-to-one orthologues
between each species pair.</li>
      <li><strong>OrthologuesStats_many-to-many.tsv</strong> contains the number of orthologues in
a many-to-many relationship for each species pair (due to gene duplication
events in both lineages post-speciation).
        <ul>
          <li>Entry (i,j) is the number of genes in species i that are in a
many-to-many orthology relationship with genes in species j.</li>
        </ul>
      </li>
      <li><strong>OrthologuesStats_one-to-many.tsv</strong>: entry (i,j) gives the number of genes in
species i that are in a one-to-many orthology relationship with genes from
species j.</li>
      <li><strong>OrthologuesStats_many-to-one.tsv</strong>: entry (i,j) gives the number of genes in
species i that are in a many-to-one orthology relationship with a gene from
species j.</li>
    </ul>
  </li>
  <li><strong>OrthologuesStats_Total.tsv</strong> contains the totals for each species pair of
orthologues of whatever multiplicity.
    <ul>
      <li>Entry (i,j) is the total number of genes in species i that have orthologues in species j.</li>
    </ul>
  </li>
  <li><strong>Statistics_Overall.tsv</strong> is a tab separated text file that contains general statistics
about orthogroup sizes and proportion of genes assigned to orthogroups.</li>
  <li><strong>Statistics_PerSpecies.tsv</strong> is a tab separated text file that contains the same
information as the Statistics_Overall.csv file but for each individual species.</li>
</ul>

<h2 id="gene_duplication_events">Gene_Duplication_Events</h2>

<ul>
  <li><strong>Duplications.tsv</strong> is a tab separated text file that lists all the gene duplication events
identified by examining each node of each orthogroup gene tree. The columns are;
  a. “Orthogroup”
  b. “Species Tree node” (see Species_Tree/SpeciesTree_rooted_node_labels.txt)
  c. “Gene tree node” (see corresponding orthogroup tree in Resolved_Gene_Trees/)
  d. “Support” (proportion of expected species for which both copies of the duplicated gene are present)
  e. “Type”
    i. “Terminal”: duplication on a terminal branch of the species tree
    ii. “Non-Terminal”: duplication on an internal branch of the species tree &amp; therefore shared by more than one species
    iii. “Non-Terminal: STRIDE”: Non-Terminal duplication that also passes the very stringent STRIDE checks for what the topology of the gene tree should be post-duplication)
  f. “Genes 1” (the list of genes descended from one of the copies of the duplicate gene)
  g. “Genes 2” (the list of genes descended from the other copy of the duplicate gene.</li>
</ul>

<p><strong>SpeciesTree_Gene_Duplications_0.5_Support.txt</strong> provides a summation of the
above duplications over the branches of the species tree. The numbers after each
node or species name are the number of gene duplication events with at least 50%
support that occurred on the branch leading to the node/species. The branch lengths
are as given in Species_Tree/SpeciesTree_rooted.txt. It is a text file, in newick
format.</p>

<h2 id="orthogroup-sequences">Orthogroup Sequences</h2>

<p>A FASTA file for each orthogroup giving the amino acid sequences for each gene in
the orthogroup.</p>

<h2 id="orthogroups-directory">Orthogroups Directory</h2>

<ul>
  <li><strong>Orthogroups.tsv</strong> is a tab separated text file that contains the genes in each orthogroup,
with columns for each species.</li>
  <li><strong>Orthogroups.txt</strong> is a text file that contains the genes in each orthogroup (one orthogroup
per line).</li>
  <li><strong>Orthogroups_UnassignedGenes.tsv</strong> is a tab separated text file that contains all of
the genes that were not assigned to any orthogroup.</li>
  <li><strong>Orthogroups.GeneCount.tsv</strong> is a tab separated text file that contains counts of the
number of genes for each species in each orthogroup.</li>
  <li><strong>Orthogroups_SingleCopyOrthologues.txt</strong> is a list of orthogroups that contain
exactly one gene per species i.e. they contain one-to-one orthologues.</li>
</ul>

<h2 id="orthologues-directory">Orthologues Directory</h2>

<p>One sub-directory for each species that in turn contains a file for each pairwise species
comparison, listing the orthologs between that species pair.</p>

<p>Orthologues can be one-to-one, one-to-many or many-to-many depending on the gene
duplication events since the orthologs diverged.</p>

<p>Each row in a file contains the gene(s) in one species that are orthologues of the gene(s) in
the other species and each row is cross-referenced to the orthogroup that contains those
genes.</p>

<h2 id="phylogenetic-hierarchical-orthogroups-directory">Phylogenetic Hierarchical Orthogroups Directory</h2>

<p>OrthoFinder infers HOGs, orthogroups at each hierarchical level (i.e. at each node in the
species tree) by analysing the rooted gene trees.</p>

<ul>
  <li><strong>N1.tsv</strong> is a tab separated text file. Each row contains the genes belonging to a single
orthogroup. The genes from each orthogroup are organized into columns, one per
species. Additional columns give the HOG (Hierarchical Orthogroup) ID and the node
in the gene tree from which the HOG was determined</li>
  <li><strong>N2.tsv, N3.tsv</strong>, …: Orthogroups inferred from the gene trees corresponding to
the clades of species in the species tree N2, N3, etc.</li>
</ul>

<h2 id="phylogenetically_misplaced_genes">Phylogenetically_Misplaced_Genes</h2>

<p>Genes in “Phylogenetically_Misplaced_Genes/” are those that appear to be out of place in
the gene tree, and would otherwise negatively affect orthology analysis if not identified.</p>

<h2 id="putative_xenologs">Putative_Xenologs</h2>

<p>Xenologs are sets of genes descended from a common ancestor, but where there has been
horizontal transfer on the evolutionary path to the gene copies in extant species, rather than
just speciation and duplication. OrthoFinder tries to identify xenologs, but we call them
‘putative’, since many arise from contamination during sequencing. Each species has a file
in this folder, listing genes and their putative xenologs from all other species</p>

<h2 id="resolved-gene-trees-directory">Resolved Gene Trees Directory</h2>

<p>A rooted phylogenetic tree inferred for each orthogroup with 4 or more sequences and
resolved using the OrthoFinder hybrid species-overlap/duplication-loss coalescent
model.</p>

<h2 id="single-copy-orthologue-sequences">Single Copy Orthologue Sequences</h2>

<p>The same files as the “Orthogroup Sequences” directory but restricted to only
those orthogroups which contain exactly one gene per species.</p>

<h2 id="species-tree-directory">Species Tree Directory</h2>

<ul>
  <li>
    <p><strong>SpeciesTree_rooted.txt</strong>: A STAG species tree inferred from all orthogroups,
containing STAG support values at internal nodes and rooted using STRIDE.</p>
  </li>
  <li>
    <p><strong>SpeciesTree_rooted_node_labels.txt</strong>: The same tree as above but with the
nodes given labels (instead of support values) to allow other results files to crossreference
branches/nodes in the species tree (e.g. location of gene duplication
events).</p>
  </li>
</ul>

<h2 id="workingdirectory">WorkingDirectory</h2>

<p>This contains all the files necessary for orthofinder to run</p>


  </div><a class="u-url" href="/OrthoFinder/tutorials/guide-to-results/" hidden></a>
</article>
<script>
    document.addEventListener("DOMContentLoaded", () => {
        fetch("https://api.github.com/repos/OrthoFinder/OrthoFinder/releases")
            .then(response => response.json())
            .then(data => {
                const assets = data.flatMap(release => release.assets);
                document.querySelectorAll(".download-count").forEach(el => {
                    const assetName = el.getAttribute("data-asset");
                    const match = assets.find(a => a.name === assetName);
                    if (match && match.download_count !== undefined) {
                        el.innerHTML = `<i class="fa fa-download" aria-hidden="true"></i> ${match.download_count} downloads`;
                    } else {
                        el.textContent = "(not found)";
                    }
                });
            })
            .catch(error => {
                console.error("GitHub API fetch error:", error);
                document.querySelectorAll(".download-count").forEach(el => {
                    el.textContent = "(error)";
                });
            });
    });
</script>
        
      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/OrthoFinder/"></data>

  <div class="wrapper">

    <div class="footer-col-wrapper">
      <div class="footer-col">

        <!-- <p class="feed-subscribe">
          <a href="http://localhost:4000/OrthoFinder/feed.xml">
            <svg class="svg-icon orange">
              <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#rss"></use>
            </svg><span>Subscribe</span>
          </a>
        </p> -->

        <!--<ul class="social-media-list"><li>
  <a rel="me" href="" target="_blank" title="GitHub">
    <svg class="svg-icon grey">
      <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
    </svg>
  </a>
</li>


</ul>
-->
         
        <div class="social-icons">
          
          <a href="https://github.com/OrthoFinder/OrthoFinder" target="_blank" title="GitHub">
            <svg class="svg-icon grey">
              <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
            </svg>
            <span>GitHub</span>
          </a>
          
          <!--  -->
        </div>


          <!--
          <ul class="contact-list">

            Link to the team section instead of listing every author
            <li>
              <a href="/OrthoFinder/about/#team">
                Meet&nbsp;the&nbsp;Team
              </a>
            </li>

          </ul> -->

          <ul class="contact-list">
            <li><a href="/OrthoFinder/about/#team">Meet the Team</a></li>
          </ul>


      </div>
      
      <div class="footer-col">
        <!-- <p>This is a software developed by the [Steven Kelly Lab](http://www.stevekellylab.com/).</p> -->
        <p><p>This is a software developed by the <a href="http://www.stevekellylab.com/">Steven Kelly Lab</a>.</p>
</p>
      </div>
    </div>

    <!-- <div class="social-links"><ul class="social-media-list"><li>
  <a rel="me" href="" target="_blank" title="GitHub">
    <svg class="svg-icon grey">
      <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
    </svg>
  </a>
</li>


</ul>
</div> -->

  </div>


<script>
  document.addEventListener('DOMContentLoaded', function () {
    // 1) Find all <pre><code> blocks
    const codeBlocks = document.querySelectorAll('pre > code');

    codeBlocks.forEach((codeBlock, index) => {
      // The parent <pre> element
      const pre = codeBlock.parentNode;

      // 2) Create a container DIV and wrap <pre> inside it
      const container = document.createElement('div');
      container.className = 'code-container';
      pre.parentNode.insertBefore(container, pre);
      container.appendChild(pre);

      // 3) Create the copy button
      const button = document.createElement('button');
      button.className = 'copy-button';
      button.type = 'button';

      // 4) Insert an inline SVG icon + "Copy" text
      //    We'll build the SVG as a string, then inject it into button.innerHTML
      const svgIcon = `
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="16"
          height="16"
          fill="currentColor"
          class="bi bi-files"
          viewBox="0 0 16 16"
        >
            <path d="M13 0H5a1 1 0 0 0-1 1v1H3a2 2 0 0 0-2
                  2v9.5A1.5 1.5 0 0 0 2.5 15H9v-1H2.5a.5.5
                  0 0 1-.5-.5V4a1 1 0 0 1 1-1h1v7.5A1.5
                  1.5 0 0 0 5.5 12H11v-1H5.5a.5.5 0 0
                  1-.5-.5V1a.5.5 0 0 1 .5-.5H5v3.5A1.5
                  1.5 0 0 0 6.5 5H13v6h1V1a1 1 0 0 0-1-1z"/>
        </svg>
    `;
      

      
      button.innerHTML = svgIcon + 'Copy';

      // 5) Assign an ID to <code> if it doesn't have one
      if (!codeBlock.id) {
        codeBlock.id = `codeblock-${index}`;
      }

      // 6) Link button to the code block by ID
      button.setAttribute('data-clipboard-target', '#' + codeBlock.id);

      // 7) Insert the button into the container
      container.insertBefore(button, pre);
    });

    // 8) Initialize Clipboard.js for all .copy-button elements
    const clipboard = new ClipboardJS('.copy-button');
    clipboard.on('success', function (e) {
      e.clearSelection();
      // Briefly change button text to indicate success
      e.trigger.innerText = 'Copied!';
      setTimeout(() => {
        e.trigger.innerHTML = svgIcon + 'Copy'; // restore icon + text
      }, 1500);
    });
    clipboard.on('error', function (e) {
      console.error('Failed to copy', e);
    });
  });
</script>


  <!-- <script>
    document.addEventListener('DOMContentLoaded', function () {
      // 1. Find all <pre><code> blocks
      const codeBlocks = document.querySelectorAll('pre > code');

      codeBlocks.forEach((codeBlock) => {
        const pre = codeBlock.parentNode;

        // 2. Create a container div to hold the code + button
        const container = document.createElement('div');
        container.className = 'code-container';

        // Insert the container before <pre>, then move <pre> inside it
        pre.parentNode.insertBefore(container, pre);
        container.appendChild(pre);

        // 3. Create a "Copy" button
        const button = document.createElement('button');
        button.className = 'copy-button';
        button.type = 'button';
        button.innerText = 'Copy';

        // If the code block doesn’t have an ID, assign one so we can target it
        let codeId = codeBlock.id;
        if (!codeId) {
          codeId = 'codeblock-' + Math.floor(Math.random() * 1000000);
          codeBlock.setAttribute('id', codeId);
        }

        // 4. Point the button at the code block
        button.setAttribute('data-clipboard-target', '#' + codeId);

        // 5. Insert button into the container
        container.insertBefore(button, pre);
      });

      // 6. Initialize Clipboard.js
      const clipboard = new ClipboardJS('.copy-button');
      clipboard.on('success', function (e) {
        e.clearSelection();
        // Change button text to confirm copy
        e.trigger.textContent = 'Copied!';
        setTimeout(() => {
          e.trigger.textContent = 'Copy';
        }, 2000);
      });
      clipboard.on('error', function (e) {
        console.error('Failed to copy', e);
      });
    });
  </script> -->


</footer></body>

</html>

