<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Tutorials | OrthoFinder</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="Tutorials" />
<meta name="author" content="David Emms, Jonathan Holmes, Laurence Belcher, Yi Liu" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="This is a software developed by the Steven Kelly Lab." />
<meta property="og:description" content="This is a software developed by the Steven Kelly Lab." />
<link rel="canonical" href="http://localhost:4000/OrthoFinder/tutorial/" />
<meta property="og:url" content="http://localhost:4000/OrthoFinder/tutorial/" />
<meta property="og:site_name" content="OrthoFinder" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Tutorials" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebPage","author":{"@type":"Person","name":"David Emms, Jonathan Holmes, Laurence Belcher, Yi Liu"},"description":"This is a software developed by the Steven Kelly Lab.","headline":"Tutorials","url":"http://localhost:4000/OrthoFinder/tutorial/"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/OrthoFinder/assets/css/style.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/OrthoFinder/feed.xml" title="OrthoFinder" /><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
</head>
<body><header class="site-header">

  <div class="wrapper"><!-- <a class="site-title" rel="author" href="/OrthoFinder/%20/">OrthoFinder</a> -->
    <a class="site-title" rel="author" href="/OrthoFinder/">OrthoFinder</a><nav class="site-nav">
      <input type="checkbox" id="nav-trigger" class="nav-trigger" />
      <label for="nav-trigger">
        <span class="menu-icon">
          <svg viewBox="0 0 18 15" width="18px" height="15px">
            <path
              d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z" />
          </svg>
        </span>
      </label>
      <div class="trigger">
        <!--<a class="page-link" href="/OrthoFinder/">Home</a><a class="page-link" href="/OrthoFinder/about/">About</a><a class="page-link" href="/OrthoFinder/download_and_install/">Download &amp; Install</a><a class="page-link" href="/OrthoFinder/tutorial/">Tutorials</a><a class="page-link" href="/OrthoFinder/faq/">FAQ</a>--><a class="page-link" href="/OrthoFinder/"><strong>Home</strong></a><a class="page-link" href="/OrthoFinder/about/"><strong>About</strong></a><a class="page-link" href="/OrthoFinder/download_and_install/"><strong>Download &amp; Install</strong></a><a class="page-link" href="/OrthoFinder/tutorial/"><strong>Tutorials</strong></a><a class="page-link" href="/OrthoFinder/faq/"><strong>FAQ</strong></a>
        <a class="page-link" href="https://github.com/OrthoFinder/OrthoFinder" target="_blank" title="GitHub Profile"><strong>GitHub</strong></a>
        
      </div>
    </nav></div>
  <script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js"></script>
</header><main class="page-content" aria-label="Content">
      <div class="wrapper">
        <div class="post">
    <!--<h1 class="page-heading">Tutorials</h1>
    <h1 class="page-heading">Download</h1>-->
    <h1 class="post-title p-name" itemprop="name headline"><strong>Tutorials</strong></h1>
    
    <div class="post-content">
        <!-- <h2 style="margin-bottom: 1.2rem;"><strong>Tutorials</strong></h2> -->
<p>Add some explanations here to direct people to the right link</p>

<!-- <ul>
  
    <li><a href="/OrthoFinder/tutorials/beginner-tutorial/">Beginner Tutorial</a></li>
  
    <li><a href="/OrthoFinder/tutorials/advanced-tutorial/">Advanced Tutorial</a></li>
  
    <li><a href="/OrthoFinder/tutorials/advanced-usage/">Advanced Usage</a></li>
  
    <li><a href="/OrthoFinder/tutorials/gettting-input-data/">Getting Input Data</a></li>
  
    <li><a href="/OrthoFinder/tutorials/guide-to-results/">Guide to Results</a></li>
  
    <li><a href="/OrthoFinder/tutorials/understand-orthology/">Understanding Orthology</a></li>
  
</ul> -->

    </div>

    <ul class="post-list"><li>
            <!-- <h2>
                <a class="post-link" href="">
                    <a href="">👉 Beginner Tutorial</a>
                </a>
            </h2> -->

            <h2>
                <a class="post-link" href="/OrthoFinder/tutorials/beginner-tutorial/">
                    <span class="post-number">1.</span>
                    Beginner Tutorial
                </a>
            </h2>

            <!-- <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i>
                

 9 minute read   
            </p> -->
            <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i> 
                

 9 minute read    | <!-- 




<span class="meta-info">
    🕒 Updated 20231 days ago

    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 20231 days ago
</span> -->







<span class="meta-info">
    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 15 days ago
</span>
            </p>

            <!-- <p class="post-meta-bar">
                <a href="/OrthoFinder/assets/docs/beginner-tutorial.pdf" target="_blank">
                    📄 Read in PDF
                </a> | 
                <a href="https://GitHub.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/beginner-tutorial.pdf" target="_blank">
                    📄 Download PDF
                </a> | 
                <script>
    document.addEventListener("DOMContentLoaded", () => {
        fetch("https://api.github.com/repos/OrthoFinder/OrthoFinder/releases")
            .then(response => response.json())
            .then(data => {
                const assets = data.flatMap(release => release.assets);
                document.querySelectorAll(".download-count").forEach(el => {
                    const assetName = el.getAttribute("data-asset");
                    const match = assets.find(a => a.name === assetName);
                    if (match && match.download_count !== undefined) {
                        el.innerHTML = `<i class="fa fa-download" aria-hidden="true"></i> ${match.download_count} downloads`;
                    } else {
                        el.textContent = "(not found)";
                    }
                });
            })
            .catch(error => {
                console.error("GitHub API fetch error:", error);
                document.querySelectorAll(".download-count").forEach(el => {
                    el.textContent = "(error)";
                });
            });
    });
</script>
            </p> -->


            <!-- <ul>
                
                <li><a href="/OrthoFinder/tutorials/beginner-tutorial/">Beginner Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-tutorial/">Advanced Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-usage/">Advanced Usage</a></li>
                
                <li><a href="/OrthoFinder/tutorials/gettting-input-data/">Getting Input Data</a></li>
                
                <li><a href="/OrthoFinder/tutorials/guide-to-results/">Guide to Results</a></li>
                
                <li><a href="/OrthoFinder/tutorials/understand-orthology/">Understanding Orthology</a></li>
                
            </ul> -->


            <!-- <span class="post-meta">May 1, 2025</span> -->

            <!-- <div class="container">
                <section id="tutorials"><p>This tutorial will cover:</p>

<ul>
  <li><a href="#downloading-orthofinder">Downloading OrthoFinder</a></li>
  <li><a href="#running-orthofinder">Running OrthoFinder</a></li>
  <li><a href="#exploring-the-results-of-orthofinder">Exploring the results of OrthoFinder</a></li>
</ul>

<p>OrthoFinder requires as input the amino acid sequences for all the protein coding genes in
your species of interest. We provide a separate tutorial for <a href="/OrthoFinder/tutorials/gettting-input-data/">getting input files</a> for OrthoFinder.</p>

<p>All these steps will be done on the command line so that you can just copy and
paste the commands yourself. If you are not familiar with the command line there
are many online tutorials and reference pages, <a href="https://www.techspot.com/guides/835-linux-command-line-basics/">here</a> is a nice short one that covers
the basics.</p>

<h3 id="downloading-orthofinder">Downloading OrthoFinder</h3>

<p>There are two main ways of getting OrthoFinder. You can either use conda, or you can
install it directly from GitHub. Installing directly from GitHub will always give you the latest
version, but you might have to manually install other software that OrthoFinder is
dependent on, and it can be trickier to troubleshoot if you aren’t familiar with the
command line. Conda automates the installation process and handles all dependencies,
making it very beginner-friendly.
To install via conda, we first need to install miniconda. Follow the instructions <a href="https://docs.anaconda.com/miniconda/">here</a>.</p>

<p>We then need to run these commands</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>conda config --add channels defaults conda config --add channels bioconda conda config
--add channels conda-forge
conda create -n orthofinder
conda activate orthofinder
conda install orthofinder
</code></pre></div></div>

<p>If you are on one of the newer Macs with the new chips (M1/M2/M3), you might need to
follow a few extra steps to use conda. Check out the guide <a href="https://towardsdatascience.com/how-to-manage-conda-environments-on-an-applesilicon-
m 1-mac-1e29cb3bad12">here</a>.</p>

<p>To install directly from GitHub, we need to run these commands</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>python3 -m venv of3_env
. of3_env/bin/activate
pip install git+https://GitHub.com/OrthoFinder/OrthoFinder.git
</code></pre></div></div>

<p>You can test that OrthoFinder has been installed by printing its help file
<code class="language-plaintext highlighter-rouge">orthofinder -h</code>, which will print all of the command line options.</p>

<p>You can test that OrthoFinder is working correctly by running it on the example
dataset, which you can download from our <a href="https://GitHub.com/OrthoFinder/OrthoFinder/">GitHub</a></p>

<p><code class="language-plaintext highlighter-rouge">orthofinder -f ExampleData/</code></p>

<p>OrthoFinder will print lots of information to the command line as it runs. If you get an
error message, the best way to troubleshoot is to just google the error message. You
can also ask a question on our <a href="https://GitHub.com/OrthoFinder/OrthoFinder/issues">GitHub</a>.</p>

<p>When OrthoFinder has finished running, it will generate a folder containing the output,
with the folder named according to today’s date.
<code class="language-plaintext highlighter-rouge">ExampleData/OrthoFinder/Results_Jul12</code> The folder looks like this:</p>

<p align="center" class="figure-wrapper">
  <img src="/OrthoFinder/assets/images/Beginner_image1.jpeg" alt="Beginner_img1" width="800" />
</p>

<p>We’ll discuss how to interpret and analyse these files and folders later on, in the
<a href="#exploring-the-results-of-orthofinder">Exploring the results</a> section of the tutorial.</p>

<h3 id="running-orthofinder">Running OrthoFinder</h3>

<p>You can now run OrthoFinder
First, you have to open a terminal and navigate to the directory where your files
are. You can now run OrthoFinder on your proteomes.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>orthofinder <span class="nt">-f</span> primary_transcripts 
</code></pre></div></div>

<p>That’s it! OrthoFinder will print updates on its progress to the terminal, and tell you when
it’s finished. To see what options you might want to adjust for your own data, check out the <a href="https://GitHub.com/OrthoFinder/OrthoFinder/">GitHub</a>, or the <a href="/OrthoFinder/tutorials/advanced-tutorial/">Advanced Tutorial</a> page</p>

<h3 id="exploring-the-results-of-orthofinder">Exploring the results of OrthoFinder</h3>

<p>OrthoFinder creates a results directory named OrthoFinder inside the proteome
directory, and puts the results here. My results directory looks like this:</p>

<p align="center" class="figure-wrapper">
  <img src="/OrthoFinder/assets/images/Beginner_image2.jpeg" alt="Beginner_img2" width="800" />
</p>

<h4 id="step-1-quality-control">Step 1: Quality Control</h4>

<p>Before we start diving into the orthogroups, it would behoove us to check the quality
of the OrthoFinder run. We want to make sure that most genes across all species
have been assigned to orthogroups, and that the species tree looks realistic.</p>

<p>Open the file Statistics_Overall.tsv from the folder <code class="language-plaintext highlighter-rouge">Comparative_Genomics_Statistics</code>. This file can be opened in spreadsheet software
like Microsoft Excel, or in a text editor like Notepad.</p>

<p>On the 5th line, we can see the ‘Percentage of genes in orthogroups’, which in my case
is <code class="language-plaintext highlighter-rouge">95.7%</code>.</p>

<p align="center" class="figure-wrapper">
  <img src="/OrthoFinder/assets/images/Beginner_image3.png" alt="Beginner_img3" width="450" />
</p>

<p>A good rule of thumb is that this number should be <code class="language-plaintext highlighter-rouge">&gt;80%</code>. If not, you are likely missing
some orthology relationships that actually exist. The best way to fix this would be better
species sampling.</p>

<p>Now open the file ‘Statistics_PerSpecies’, from the same folder. This file gives us the
<code class="language-plaintext highlighter-rouge">%</code> of genes in each species that are assigned to orthogroups, rather than the
percentage for all genes across species.</p>

<p>You can see here that we capture most genes across all species.</p>

<p align="center" class="figure-wrapper">
  <img src="/OrthoFinder/assets/images/Beginner_image4.png" alt="Beginner_img4" width="1000" />
</p>

<p>The lowest percentage is the kiwi (<em><code class="language-plaintext highlighter-rouge">A. haastii</code></em>), but we still managed to assign <code class="language-plaintext highlighter-rouge">94%</code> of its
genes to orthogroups. The key message here is that it’s always a good idea to look at
this information before you start interpreting your results. If the numbers were too low for
one species, we might want to consider sampling more species to fill in the long
evolutionary divergence between species (e.g. something in between a Kiwi and a
Kakapo, such as a Hoatzin).</p>

<p>One more useful thing to do before we really start to dive in is to look at the species
tree. Go to the <a href="https://itol.embl.de/">iTOL website</a>, and click <code class="language-plaintext highlighter-rouge">Upload a tree</code>.
You can then drag and drop the tree file, which is in <code class="language-plaintext highlighter-rouge">Species_Tree/SpeciesTree_rooted.txt</code>
You will now see the phylogenetic tree that OrthoFinder has produced. I have annotated
my version with icons PhyloPic, so that we can see what is going on</p>

<p align="center" class="figure-wrapper">
  <img src="/OrthoFinder/assets/images/Beginner_image5.png" alt="Beginner_img5" width="800" />
</p>

<p>We now want to do some common-sense checking that everything appears to be in
order, and we aren’t rewriting the history of life on earth. With our six species, this tree
looks exactly as we would expect.
If the tree doesn’t look correct, then this won’t impact orthogroup inference, but will affect
our measures of gene duplication, and might affect our assignment of orthologs and
paralogs within an orthogroup. If you need to, you can run OrthoFinder with your own
species tree (use the -s option).</p>

<h4 id="step-2-interpreting-results">Step 2: Interpreting results</h4>

<p>Now that we are happy with our OrthoFinder run, we can start diving into the results.</p>

<ul>
  <li>
    <p><strong><em>Orthologues</em></strong><br />
  We will start by finding orthologues of a gene that we are interested in. We will focus on the gene <code class="language-plaintext highlighter-rouge">ENSVURG00010002700.1</code> in wombats, which is an
  olfactory receptor. Let’s find out what its orthologues are in the Tammar wallaby.
  In the Orthologues directory there is a sub-directory for each species.</p>

    <p>Open <code class="language-plaintext highlighter-rouge">Orthologues/Orthologues_Vombatus_ursinus/Vombatus_ursinus v
  Notamacropus_eugeni i.tsv</code>, in a spreadsheet program (specifying that it’s tab-delimited
  if necessary). The file has three columns, <code class="language-plaintext highlighter-rouge">Orthogroup</code>, <code class="language-plaintext highlighter-rouge">Vombatus_ursinus</code>, and
  <code class="language-plaintext highlighter-rouge">Notamacropus_eugenii</code>. Find <code class="language-plaintext highlighter-rouge">ENSVURG00010002700.1</code> in the table, I can see that
  the gene is in orthogroup <code class="language-plaintext highlighter-rouge">OG0000365</code> and that it has three orthologues in wallabies:
  <code class="language-plaintext highlighter-rouge">ENSMEUG00000016785.1</code>, <code class="language-plaintext highlighter-rouge">ENSMEUG00000001827.1</code>, <code class="language-plaintext highlighter-rouge">ENSMEUG00000010434.1</code></p>
  </li>
  <li>
    <p><strong><em>Gene trees</em></strong><br />
  Next, we are going to look at the gene tree to see how these orthologues arose.
  OrthoFinder infers orthlologues from <code class="language-plaintext highlighter-rouge">resolved</code> gene trees using a Duplication-Loss-
  Coalescence analysis to identify the more parsimonious interpretation of the tree (see
  the OrthoFinder2 <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1832-y">paper</a> for more details).</p>

    <p>All of the gene trees are in one file (<code class="language-plaintext highlighter-rouge">Resolved_Gene_Trees/Resolved_Gene_Trees.txt</code>).
  Each line of the file contains the ID of an orthogroup (e.g. <code class="language-plaintext highlighter-rouge">OG0000365</code>:), followed by the
  gene tree for that orthogroup. To find the tree for certain orthogroup, just search for the
  orthogroup ID.</p>

    <p>We are going to view the tree for <code class="language-plaintext highlighter-rouge">OG0000365</code> on <a href="https://itol.embl.de/">iTOL</a>.</p>

    <p align="center" class="figure-wrapper">
  <img src="/OrthoFinder/assets/images/Beginner_image6.png" alt="Beginner_img6" width="1000" />
  </p>

    <p>Looking at the gene tree, we can see that there have been several gene duplications in
  the lineage leading to wallabies (<em><code class="language-plaintext highlighter-rouge">Notamacropus</code></em>). This has resulted in a one-to-three
  orthology relationship, i.e. all three of the wallaby genes are equally related to the
  wombat gene <code class="language-plaintext highlighter-rouge">ENSVURG00010002700.1</code>. It’s often the case that orthology relationships
  aren’t one-to-one, and it’s important to know this—you don’t want to spend months doing
  experiments on ‘the orthologue’ only to find out later there are actually three!</p>
  </li>
  <li>
    <p><strong><em>Gene duplications</em></strong><br />
  Having the gene trees means that OrthoFinder can identify all gene duplication events
  that occurred. There is a folder called <code class="language-plaintext highlighter-rouge">Gene_Duplication_Events</code> that has two files that
  allow us to explore duplications. Let’s first open
  <code class="language-plaintext highlighter-rouge">Gene_Duplication_Events/SpeciesTree_Gene_Duplications_0.5_Support.txt</code> in iTOL.
  Go into the <code class="language-plaintext highlighter-rouge">Advanced</code> tab on the Control Panel and select <code class="language-plaintext highlighter-rouge">Display</code> next to <code class="language-plaintext highlighter-rouge">Node IDs</code>
  to see the node labels</p>

    <p align="center" class="figure-wrapper">
  <img src="/OrthoFinder/assets/images/Beginner_image7.png" alt="Beginner_img7" width="1200" />
  </p>

    <p>This gives a summary of gene duplication events. Each node shows the node name
  followed by an underscore and then the number of well-supported gene duplication
  events mapped to each node in the species tree. Gene-duplication events are
  considered <code class="language-plaintext highlighter-rouge">well-supported</code> if at least <code class="language-plaintext highlighter-rouge">50%</code> of the descendant species have retained
  both copies of the duplicated gene. For the common ancestor of the mammals, N2, there were 812 of these well-supported gene duplication events. The numbers after the
  species names are the number of <code class="language-plaintext highlighter-rouge">terminal</code> duplications that map to that species, rather
  than an internal node of the species tree.</p>

    <p>We can see the full list of gene duplication
  events in the file <code class="language-plaintext highlighter-rouge">Gene_Duplication_Events/Duplications.tsv</code>. Here are just a few lines
  from the file:</p>

    <p align="center" class="figure-wrapper">
  <img src="/OrthoFinder/assets/images/Beginner_image8.png" alt="Beginner_img8" width="800" />
  </p>

    <p>Each gene duplication event is cross-referenced to the species tree node, and the node
  in the gene tree. It also lists the genes descended from each of the two copies arising
  from the gene duplication event. We can check this out for our wombat olfactory
  receptor orthologues.</p>

    <p align="center" class="figure-wrapper">
  <img src="/OrthoFinder/assets/images/Beginner_image9.png" alt="Beginner_img9" width="1200" />
  </p>

    <p>These events are also summarised by orthogroup and by species tree node in the files
  Duplications_per_Orthogroup.tsv and Duplications_per_Species_Tree_Node.tsv which are
  both in the directory <code class="language-plaintext highlighter-rouge">Comparative_Genomics_Statistics/</code>.</p>
  </li>
  <li>
    <p><strong><em>Orthogroups</em></strong><br />
  Often we’re interested in group-wise species comparisons, that is comparisons across a
  clade of species rather than between a pair of species. The generalisation of orthology
  to multiple species is the orthogroup. Just like orthologues are the genes descended
  from a single gene in the last common ancestor of a pair of species an orthogroup is the
  set of genes descended from a single gene in a group of species. Each gene tree from
  OrthoFinder, for example the one above, is for one orthogroup. The orthogroup gene
  tree is the tree we need to look at if we want it to include all pairwise orthologues. And
  even though some of the genes within an orthogroup can be paralogs of one another, if
  we tried to take any genes out then we would also be removing orthologs too.</p>

    <p>So if we want to do a comparison of the <code class="language-plaintext highlighter-rouge">equivalent</code> genes in a set of species, we need
  to do the comparison across the genes in an orthogroup. The orthogroups are in the file
  <code class="language-plaintext highlighter-rouge">Orthogroups/Orthogroups.tsv</code>. This table has one orthogroup per line and one species
  per column and is ordered from largest orthogroup to smallest.</p>
  </li>
  <li>
    <p><strong><em>Hierarchical Orthogroups</em></strong><br />
  OrthoFinder3 also infers hierarchical orthogroups for each node in the species tree. A
  file equivalent to <code class="language-plaintext highlighter-rouge">Orthogroups/Orthogroups.tsv</code> is available for each node in
  <code class="language-plaintext highlighter-rouge">/Phylogenetic_Hierarchical_Orthogroups</code>. You can compare the node number (e.g. <code class="language-plaintext highlighter-rouge">N3</code>)
  to the species tree, to see which species will be included.</p>
  </li>
  <li>
    <p><strong><em>Orthogroup sequences</em></strong><br />
  For each orthogroup there is a FASTA file in Orthogroup_Sequences/ which contains the
  sequences for the genes in that orthogroup.</p>
  </li>
  <li>
    <p><strong><em>Other results files</em></strong><br />
  We have now covered all of the main output files that will be useful to most users, but
  OrthoFinder also outputs much more useful information! A full description of the output
  files is available below.</p>

    <p>There are also some useful community tools that allow interactive viewing of results,
  such as <a href="https://orthobrowserexamples.netlify.app/">OrthoBrowser</a></p>
  </li>
</ul>

</section>
            </div> -->
        </li><li>
            <!-- <h2>
                <a class="post-link" href="">
                    <a href="">👉 Advanced Tutorial</a>
                </a>
            </h2> -->

            <h2>
                <a class="post-link" href="/OrthoFinder/tutorials/advanced-tutorial/">
                    <span class="post-number">2.</span>
                    Advanced Tutorial
                </a>
            </h2>

            <!-- <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i>
                

 1 minute read   
            </p> -->
            <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i> 
                

 1 minute read    | <!-- 




<span class="meta-info">
    🕒 Updated 20231 days ago

    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 20231 days ago
</span> -->







<span class="meta-info">
    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 16 days ago
</span>
            </p>

            <!-- <p class="post-meta-bar">
                <a href="/OrthoFinder/assets/docs/advanced-usage.pdf" target="_blank">
                    📄 Read in PDF
                </a> | 
                <a href="https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/advanced-usage.pdf" target="_blank">
                    📄 Download PDF
                </a> | 
                <script>
    document.addEventListener("DOMContentLoaded", () => {
        fetch("https://api.github.com/repos/OrthoFinder/OrthoFinder/releases")
            .then(response => response.json())
            .then(data => {
                const assets = data.flatMap(release => release.assets);
                document.querySelectorAll(".download-count").forEach(el => {
                    const assetName = el.getAttribute("data-asset");
                    const match = assets.find(a => a.name === assetName);
                    if (match && match.download_count !== undefined) {
                        el.innerHTML = `<i class="fa fa-download" aria-hidden="true"></i> ${match.download_count} downloads`;
                    } else {
                        el.textContent = "(not found)";
                    }
                });
            })
            .catch(error => {
                console.error("GitHub API fetch error:", error);
                document.querySelectorAll(".download-count").forEach(el => {
                    el.textContent = "(error)";
                });
            });
    });
</script>
            </p> -->


            <!-- <ul>
                
                <li><a href="/OrthoFinder/tutorials/beginner-tutorial/">Beginner Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-tutorial/">Advanced Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-usage/">Advanced Usage</a></li>
                
                <li><a href="/OrthoFinder/tutorials/gettting-input-data/">Getting Input Data</a></li>
                
                <li><a href="/OrthoFinder/tutorials/guide-to-results/">Guide to Results</a></li>
                
                <li><a href="/OrthoFinder/tutorials/understand-orthology/">Understanding Orthology</a></li>
                
            </ul> -->


            <!-- <span class="post-meta">May 1, 2025</span> -->

            <!-- <div class="container">
                <section id="tutorials"><p>OrthoFinder3 provides a new workflow to assign new genes from new species to an already 
inferred set of orthogroups for a smaller, core group of species. If you are analysing <code class="language-plaintext highlighter-rouge">&gt;100</code> species, we reccommend that you use this scalable implementation.</p>

<p>To do this effectively, we need to pick a good set of core species. If we are running
OrthoFinder on some bees, some moths, and some flies, we want to build our core 
orthogroups on a phylogenetically diverse set of those species. If instead we chose 
randomly and ended up using all moths as our core species, we might end up with less 
accurate orthogroups, and a longer runtime.</p>

<p>If you have a phylogenetic tree, you can use tools like Phylogenetic Diversity Analyzer to pick a representative set (http://www.cibiv.at/software/pda/). Or, you could just pick a diverse set yourself. We recommend that you pick 64 species for your core set. 
First, run OrthoFinder on the subset of 64 species</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>orthofinder <span class="o">[</span>options] <span class="nt">-f</span> &lt;core&gt;
</code></pre></div></div>

<p>Then, add the additional species to the results of the core run</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>orthofinder <span class="o">[</span>options] <span class="nt">--assign</span> &lt;additional&gt; <span class="nt">--core</span> &lt;Results_Dir&gt;
</code></pre></div></div>
<p>This workflow also makes it really quick and easy to add new species to previous 
OrthoFinder runs. For example, if your research group works on various species of 
Angiosperms you might collectively share a core OrthoFinder results folder with a 
phylogenetically diverse set of species, which individual researchers could easily add any new species to.</p>

<p>An important note is that this workflow requires multiple sequence alignments, so 
unfortunately you cannot use it to add to OrthoFinder2 results that were run with the
default <code class="language-plaintext highlighter-rouge">-M</code> dendroblast option.</p>

<p>You can make orthogroup inference more accurate by including <strong>outgroup species</strong>. You 
just need to make sure that you use the correct <code class="language-plaintext highlighter-rouge">N*.tsv</code> file in 
Phylogenetic_Hierarchical_Orthogroups to look at the orthogroups (use species tree to 
discover which one)</p>
</section>
            </div> -->
        </li><li>
            <!-- <h2>
                <a class="post-link" href="">
                    <a href="">👉 Advanced Usage</a>
                </a>
            </h2> -->

            <h2>
                <a class="post-link" href="/OrthoFinder/tutorials/advanced-usage/">
                    <span class="post-number">3.</span>
                    Advanced Usage
                </a>
            </h2>

            <!-- <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i>
                

 1 minute read   
            </p> -->
            <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i> 
                

 1 minute read    | <!-- 




<span class="meta-info">
    🕒 Updated 20231 days ago

    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 20231 days ago
</span> -->







<span class="meta-info">
    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 17 days ago
</span>
            </p>

            <!-- <p class="post-meta-bar">
                <a href="/OrthoFinder/assets/docs/advanced-usage.pdf" target="_blank">
                    📄 Read in PDF
                </a> | 
                <a href="https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/advanced-usage.pdf" target="_blank">
                    📄 Download PDF
                </a> | 
                <script>
    document.addEventListener("DOMContentLoaded", () => {
        fetch("https://api.github.com/repos/OrthoFinder/OrthoFinder/releases")
            .then(response => response.json())
            .then(data => {
                const assets = data.flatMap(release => release.assets);
                document.querySelectorAll(".download-count").forEach(el => {
                    const assetName = el.getAttribute("data-asset");
                    const match = assets.find(a => a.name === assetName);
                    if (match && match.download_count !== undefined) {
                        el.innerHTML = `<i class="fa fa-download" aria-hidden="true"></i> ${match.download_count} downloads`;
                    } else {
                        el.textContent = "(not found)";
                    }
                });
            })
            .catch(error => {
                console.error("GitHub API fetch error:", error);
                document.querySelectorAll(".download-count").forEach(el => {
                    el.textContent = "(error)";
                });
            });
    });
</script>
            </p> -->


            <!-- <ul>
                
                <li><a href="/OrthoFinder/tutorials/beginner-tutorial/">Beginner Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-tutorial/">Advanced Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-usage/">Advanced Usage</a></li>
                
                <li><a href="/OrthoFinder/tutorials/gettting-input-data/">Getting Input Data</a></li>
                
                <li><a href="/OrthoFinder/tutorials/guide-to-results/">Guide to Results</a></li>
                
                <li><a href="/OrthoFinder/tutorials/understand-orthology/">Understanding Orthology</a></li>
                
            </ul> -->


            <!-- <span class="post-meta">May 1, 2025</span> -->

            <!-- <div class="container">
                <section id="tutorials"><h3 id="manually-installing-dependencies">Manually installing dependencies</h3>

<p>To run OrthoFinder3 on default setting, you will need diamond, famsa, fasttree, MCL
If you install OrthoFinder using a recommended method, you shouldn’t need to install
any of the dependencies manually, but in case you do;</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># Diamond</span>
https://github.com/bbuchfink/diamond
<span class="c"># FAMSA</span>
https://github.com/refresh-bio/FAMSA
<span class="c"># FastTree</span>
http://www.microbesonline.org/fasttree/#Install
<span class="c"># MCL</span>
https://github.com/micans/mcl
</code></pre></div></div>
<h3 id="running-blast-searches-separately">Running BLAST searches separately</h3>
<p>The <code class="language-plaintext highlighter-rouge">-op</code> option will prepare the files in the format required by
OrthoFinder and print the set of BLAST commands to run.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>orthofinder <span class="nt">-f</span> fasta_files_directory <span class="nt">-op</span>
</code></pre></div></div>
<p>This is useful if you want to manage the BLAST searches yourself. For example, you
may want to distribute them across multiple machines. Once the BLAST searches have 
been completed the orthogroups can be calculated using the <code class="language-plaintext highlighter-rouge">-b</code> command (see below)</p>

<h3 id="use-pre-computed-blast">Use pre-computed BLAST</h3>
<p>It is possible to run OrthoFinder with pre-computed BLAST results provided they are in 
the correct format. They can be prepared in the correct format using the <code class="language-plaintext highlighter-rouge">-op</code> command 
and, equally, the files from a previous OrthoFinder run are also in the correct format to rerun using the <code class="language-plaintext highlighter-rouge">-b</code> option. The command is simply:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>orthofinder <span class="nt">-b</span> directory_with_processed_fasta_and_blast_results
</code></pre></div></div>
<p>If you are running the BLAST searches yourself it is strongly recommended that you
use the <code class="language-plaintext highlighter-rouge">-op</code> option to prepare the files first (see Section <a href="#running-blast-searches-separately">Running BLAST Searches 
Separately</a>)</p>
</section>
            </div> -->
        </li><li>
            <!-- <h2>
                <a class="post-link" href="">
                    <a href="">👉 Getting Input Data</a>
                </a>
            </h2> -->

            <h2>
                <a class="post-link" href="/OrthoFinder/tutorials/gettting-input-data/">
                    <span class="post-number">4.</span>
                    Getting Input Data
                </a>
            </h2>

            <!-- <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i>
                

 5 minute read   
            </p> -->
            <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i> 
                

 5 minute read    | <!-- 




<span class="meta-info">
    🕒 Updated 20231 days ago

    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 20231 days ago
</span> -->







<span class="meta-info">
    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 17 days ago
</span>
            </p>

            <!-- <p class="post-meta-bar">
                <a href="/OrthoFinder/assets/docs/gettting-input-data.pdf" target="_blank">
                    📄 Read in PDF
                </a> | 
                <a href="https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/gettting-input-data.pdf" target="_blank">
                    📄 Download PDF
                </a> | 
                <script>
    document.addEventListener("DOMContentLoaded", () => {
        fetch("https://api.github.com/repos/OrthoFinder/OrthoFinder/releases")
            .then(response => response.json())
            .then(data => {
                const assets = data.flatMap(release => release.assets);
                document.querySelectorAll(".download-count").forEach(el => {
                    const assetName = el.getAttribute("data-asset");
                    const match = assets.find(a => a.name === assetName);
                    if (match && match.download_count !== undefined) {
                        el.innerHTML = `<i class="fa fa-download" aria-hidden="true"></i> ${match.download_count} downloads`;
                    } else {
                        el.textContent = "(not found)";
                    }
                });
            })
            .catch(error => {
                console.error("GitHub API fetch error:", error);
                document.querySelectorAll(".download-count").forEach(el => {
                    el.textContent = "(error)";
                });
            });
    });
</script>
            </p> -->


            <!-- <ul>
                
                <li><a href="/OrthoFinder/tutorials/beginner-tutorial/">Beginner Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-tutorial/">Advanced Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-usage/">Advanced Usage</a></li>
                
                <li><a href="/OrthoFinder/tutorials/gettting-input-data/">Getting Input Data</a></li>
                
                <li><a href="/OrthoFinder/tutorials/guide-to-results/">Guide to Results</a></li>
                
                <li><a href="/OrthoFinder/tutorials/understand-orthology/">Understanding Orthology</a></li>
                
            </ul> -->


            <!-- <span class="post-meta">May 1, 2025</span> -->

            <!-- <div class="container">
                <section id="tutorials"><h3 id="downloading-input-files-for-orthofinder">Downloading input files for OrthoFinder</h3>

<p>Let’s imagine we want to perform a phylogenomic analysis across a set of species:
Dingo (<em><code class="language-plaintext highlighter-rouge">Canis lupus dingo</code></em>), Great spotted kiwi (<em><code class="language-plaintext highlighter-rouge">Apteryx haastii</code></em>), Kakapo (<em><code class="language-plaintext highlighter-rouge">Strigops
habroptila</code></em>), Platypus (<em><code class="language-plaintext highlighter-rouge">Ornithorhynchus anatinus</code></em>), Tammar wallaby (<em><code class="language-plaintext highlighter-rouge">Notamacropus
eugenii</code></em>), and Common wombat (<em><code class="language-plaintext highlighter-rouge">Vombatus ursinus</code></em>).</p>

<p>OrthoFinder requires as input the amino acid sequences for all the protein coding genes
in your species of interest. In this section of the tutorial we will discuss how to get these
files for the species that you want to analyse. We will cover three major websites for
getting this data, Ensembl, NCBI, and Phytozome.</p>

<ul>
  <li>
    <p><a href="#getting-data-from-ensembl"><strong>Ensembl</strong></a><br />
Recommended if you want to analyse eukaryotic species, especially
vertebrates and model organisms.</p>
  </li>
  <li>
    <p><a href="#getting-data-from-ncbi"><strong>NCBI</strong></a><br />
Recommended if you want to analyse prokaryotic species.</p>
  </li>
  <li>
    <p><a href="#getting-data-from-phytozome"><strong>Phytozome</strong></a><br />
Recommended if you want to analyse plants.</p>
  </li>
</ul>

<p>A key consideration when getting input data for OrthoFinder is gene transcripts. When you download the amino acid sequences for a genome, you might have multiple
sequences for the same gene. This is because a single gene can produce multiple
transcripts, which each might produce a different protein. You can see an example of
transcripts in this table of the Actin gene on <a href="http://www.ensembl.org/Homo_sapiens/Gene/Splice?db=core;g=ENSG00000075624;r=7:5527151-5563784">Ensembl</a>.</p>

<p>If we ran OrthoFinder on these raw files it would take much longer than necessary, and
could lower the accuracy. We therefore want to extract just the longest transcript variant for
each gene. OrthoFinder provides scripts to do this, which we will learn how to use later.</p>

<h3 id="getting-data-from-ensembl">Getting data from Ensembl</h3>

<p>First, we go to the <a herf="https://www.ensembl.org/info/about/species.html">Ensembl webpage</a> with the list of species. You can also use the new <a herf="https://beta.ensembl.org/species-selector">beta website</a>, which is updated more regularly.</p>

<p>We can then search for our first species (<em><code class="language-plaintext highlighter-rouge">Canis lupus dingo</code></em>), and click on the
link This will take us to the ensembl page for that species</p>

<p align="center">
  <img src="/OrthoFinder/assets/images/GettingInput_image1.jpeg" alt="input_img1" width="1000" />
</p>

<p>On the right hand side under the heading <code class="language-plaintext highlighter-rouge">Gene annotation</code> we can press the link to
<code class="language-plaintext highlighter-rouge">Download FASTA</code> files for the genome.</p>

<p>This will take us to a webpage with some folders.</p>

<p align="center">
  <img src="/OrthoFinder/assets/images/GettingInput_image2.png" alt="input_img2" width="600" />
</p>

<p>We need to click on the <code class="language-plaintext highlighter-rouge">pep/</code> folder, and then click on the file that ends in .pep.all.fa.gz</p>

<p align="center">
  <img src="/OrthoFinder/assets/images/GettingInput_image3.jpeg" alt="input_img3" width="600" />
</p>

<p>We can then repeat this step for our other species, and place the files that are
downloaded into a folder on our computer. I have named my folder <code class="language-plaintext highlighter-rouge">Proteomes</code>.</p>

<p>The files are stored as .gz compressed files to save space, but we now need to expand
the files so that we can access the sequences. You can either double-click on them all,
or use the command line</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">gunzip</span> <span class="k">*</span>.gz
</code></pre></div></div>

<p>We’ll use a script provided with OrthoFinder to extract just the longest transcript variant
per gene and run OrthoFinder on these files:</p>

<p>You can find the script <a href="https://github.com/OrthoFinder/OrthoFinder/blob/master/tools/primary_transcript.py">here</a></p>

<p>To run the script, first place it in the Proteomes folder. Then, open the command line
and navigate to the Proteomes folder (using cd). You can then use the following
command to run the script</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for </span>f <span class="k">in</span> <span class="k">*</span>fa <span class="p">;</span> <span class="k">do </span>python primary_transcript.py <span class="nv">$f</span> <span class="p">;</span> <span class="k">done</span>
</code></pre></div></div>

<p>The script will generate a new folder called <code class="language-plaintext highlighter-rouge">primary_transcripts</code>, which contains our files.
Shortening the filenames is a good idea as it keeps the results tidy as the filenames are
used to refer to the species, e.g. I shortened it to Canis_lupus_dingo.fa.</p>

<p>Our data is now ready for OrthoFinder. You can go to the ‘Beginner Tutorial’ to see how to run OrthoFinder,
or check out the other guides on getting input data from other sources</p>

<h3 id="getting-data-from-ncbi">Getting data from NCBI</h3>

<p>We start by going to https://www.ncbi.nlm.nih.gov/ and searching for our first species</p>

<p align="center">
  <img src="/OrthoFinder/assets/images/GettingInput_image4.png" alt="input_img4" width="1000" />
</p>

<p>We want to find the genome of this species, so scroll down to the <code class="language-plaintext highlighter-rouge">Genomes</code> section,
and click on <code class="language-plaintext highlighter-rouge">Assembly / Genome</code></p>

<p align="center">
  <img src="/OrthoFinder/assets/images/GettingInput_image5.jpeg" alt="input_img5" width="1000" />
</p>

<p>There might be several genomes listed on this page. We are going to click on the one
with the green tick, which shows us the reference genome.</p>

<p align="center">
  <img src="/OrthoFinder/assets/images/GettingInput_image6.png" alt="input_img6" width="1000" />
</p>

<p>We can then click the blue <code class="language-plaintext highlighter-rouge">Download</code> button, and click the boxes to select the curated
<code class="language-plaintext highlighter-rouge">RefSeq only</code> annotation, and that we want the GFF and protein FASTA files. We can
also give our download a useful name (such as the name of the species).</p>

<p align="center">
  <img src="/OrthoFinder/assets/images/GettingInput_image7.jpeg" alt="input_img7" width="500" />
</p>

<p>We can then click the <code class="language-plaintext highlighter-rouge">Download</code> button, which downloads a zip file.</p>

<p>Repeat this step for all of the species that you want to use. For some species in our
example list, like the Kiwi, there is no RefSeq genome with protein fasta available. For
these species, it is recommended to use Ensembl.</p>

<p>Now, we need to deal with the issue of multiple transcripts per gene. For our first
species, the Dingo, there are about 20,000 genes. However, if we look at the
<code class="language-plaintext highlighter-rouge">protein.faa</code> file that we have downloaded, it has about 75,000 sequences.</p>

<p>We can use the script <code class="language-plaintext highlighter-rouge">ncbi_primary_transcript.py</code> to extract the longest
transcript per gene.</p>

<p>Place the script in the folder that has the zip folders, and run the following line of code in the
command line</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>python ncbi_primary_transcripts.py
</code></pre></div></div>

<p>This will then make a folder named <code class="language-plaintext highlighter-rouge">primary_transcripts</code>, which is ready to run
Orthofinder on.</p>

<p>If you want to download data in bulk, you can also use the NCBI Datasets tool, which
can run on the command line. For more info, see <a href="https://www.ncbi.nlm.nih.gov/datasets/">here</a></p>

<h3 id="getting-data-from-phytozome">Getting data from Phytozome</h3>

<p>First, go to the <a href="https://data.jgi.doe.gov/">Phytozome data portal website</a>. Here, you can search for a species, such
as Oryza sativa</p>

<p align="center">
  <img src="/OrthoFinder/assets/images/GettingInput_image8.jpeg" alt="input_img8" width="800" />
</p>

<p>You can then find the genome that want to download, and click on it</p>

<p>The file that we want ends in <code class="language-plaintext highlighter-rouge">primaryTranscriptOnly.fa.gz</code>. Luckily for us, Phytozome
already deals with the problem of multiple transcripts. We can select that file, and click
<code class="language-plaintext highlighter-rouge">Add to Cart</code>.</p>

<p align="center">
  <img src="/OrthoFinder/assets/images/GettingInput_image9.jpeg" alt="input_img9" width="1000" />
</p>

<p>We can then repeat this step for all of the species that we want to download. When we
are ready, we then click the shopping cart icon, and press <code class="language-plaintext highlighter-rouge">Download</code> on the next page.
You will have to sign-in or register to download data.</p>

<p>These files can be placed in a folder, where they are now ready to run OrthoFinder.</p>

</section>
            </div> -->
        </li><li>
            <!-- <h2>
                <a class="post-link" href="">
                    <a href="">👉 Guide to Results</a>
                </a>
            </h2> -->

            <h2>
                <a class="post-link" href="/OrthoFinder/tutorials/guide-to-results/">
                    <span class="post-number">5.</span>
                    Guide to Results
                </a>
            </h2>

            <!-- <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i>
                

 5 minute read   
            </p> -->
            <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i> 
                

 5 minute read    | <!-- 




<span class="meta-info">
    🕒 Updated 20231 days ago

    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 20231 days ago
</span> -->







<span class="meta-info">
    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 17 days ago
</span>
            </p>

            <!-- <p class="post-meta-bar">
                <a href="/OrthoFinder/assets/docs/guide-to-results.pdf" target="_blank">
                    📄 Read in PDF
                </a> | 
                <a href="https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/guide-to-results.pdf" target="_blank">
                    📄 Download PDF
                </a> | 
                <script>
    document.addEventListener("DOMContentLoaded", () => {
        fetch("https://api.github.com/repos/OrthoFinder/OrthoFinder/releases")
            .then(response => response.json())
            .then(data => {
                const assets = data.flatMap(release => release.assets);
                document.querySelectorAll(".download-count").forEach(el => {
                    const assetName = el.getAttribute("data-asset");
                    const match = assets.find(a => a.name === assetName);
                    if (match && match.download_count !== undefined) {
                        el.innerHTML = `<i class="fa fa-download" aria-hidden="true"></i> ${match.download_count} downloads`;
                    } else {
                        el.textContent = "(not found)";
                    }
                });
            })
            .catch(error => {
                console.error("GitHub API fetch error:", error);
                document.querySelectorAll(".download-count").forEach(el => {
                    el.textContent = "(error)";
                });
            });
    });
</script>
            </p> -->


            <!-- <ul>
                
                <li><a href="/OrthoFinder/tutorials/beginner-tutorial/">Beginner Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-tutorial/">Advanced Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-usage/">Advanced Usage</a></li>
                
                <li><a href="/OrthoFinder/tutorials/gettting-input-data/">Getting Input Data</a></li>
                
                <li><a href="/OrthoFinder/tutorials/guide-to-results/">Guide to Results</a></li>
                
                <li><a href="/OrthoFinder/tutorials/understand-orthology/">Understanding Orthology</a></li>
                
            </ul> -->


            <!-- <span class="post-meta">May 1, 2025</span> -->

            <!-- <div class="container">
                <section id="tutorials"><p><em>Definitions for some terms that are used in these files:</em></p>

<ul>
  <li><strong>Species-specific orthogroup</strong>: An orthogroups that consist entirely of genes from
one species.</li>
  <li><strong>G50</strong>: The number of genes in the orthogroup such that 50% of genes are in
orthogroups of that size or larger.</li>
  <li><strong>O50</strong>: The smallest number of orthogroups such that 50% of genes are in orthogroups
of that size or larger.</li>
  <li><strong>Single-copy orthogroup</strong>: An orthogroup with exactly one gene (and no more) from
each species.</li>
  <li><strong>Unassigned gene</strong>: A gene that has not been put into an orthogroup with any other
genes.</li>
</ul>

<h3 id="comparative_genomics_statistics">Comparative_Genomics_Statistics</h3>

<ul>
  <li>
    <p><strong>Duplications_per_Orthogroup.tsv</strong><br /> 
A tab separated text file that gives the number of duplications identified in each orthogroup.</p>
  </li>
  <li>
    <p><strong>Duplications_per_Species_Tree_Node.tsv</strong><br />
A tab separated text file that gives the
number of duplications identified as occurring along each branch of the species tree.</p>
  </li>
  <li>
    <p><strong>Orthogroups_SpeciesOverlaps.tsv</strong><br />
A tab separated text file that contains the number of orthogroups shared between each species-pair as a square matrix.</p>
  </li>
  <li>
    <p><strong>OrthologuesStats_*.tsv</strong><br /> 
These files are tab separated text files containing matrices giving the numbers of orthologues in one-to-one, one-to-many and many-to-many relationships between each pair of species.</p>

    <ul>
      <li>
        <p><strong>OrthologuesStats_one-to-one.tsv</strong><br /> 
It is the number of one-to-one orthologues between each species pair.</p>
      </li>
      <li>
        <p><strong>OrthologuesStats_many-to-many.tsv</strong><br />
It contains the number of orthologues in
a many-to-many relationship for each species pair (due to gene duplication
events in both lineages post-speciation).</p>

        <p>Entry <code class="language-plaintext highlighter-rouge">(i,j)</code> is the number of genes in species <code class="language-plaintext highlighter-rouge">i</code> that are in a
many-to-many orthology relationship with genes in species <code class="language-plaintext highlighter-rouge">j</code>.</p>
      </li>
      <li>
        <p><strong>OrthologuesStats_one-to-many.tsv</strong><br />
Entry <code class="language-plaintext highlighter-rouge">(i,j)</code> gives the number of genes in
species <code class="language-plaintext highlighter-rouge">i</code> that are in a one-to-many orthology relationship with genes from
species <code class="language-plaintext highlighter-rouge">j</code>.</p>
      </li>
      <li>
        <p><strong>OrthologuesStats_many-to-one.tsv</strong><br />
Entry <code class="language-plaintext highlighter-rouge">(i,j)</code> gives the number of genes in
species <code class="language-plaintext highlighter-rouge">i</code> that are in a many-to-one orthology relationship with a gene from
species <code class="language-plaintext highlighter-rouge">j</code>.</p>
      </li>
    </ul>
  </li>
  <li>
    <p><strong>OrthologuesStats_Total.tsv</strong><br /> 
It contains the totals for each species pair of
orthologues of whatever multiplicity.</p>

    <p>Entry <code class="language-plaintext highlighter-rouge">(i,j)</code> is the total number of genes in species <code class="language-plaintext highlighter-rouge">i</code> that have orthologues in species <code class="language-plaintext highlighter-rouge">j</code>.</p>
  </li>
  <li>
    <p><strong>Statistics_Overall.tsv</strong><br /> 
A tab separated text file that contains general statistics about orthogroup sizes and proportion of genes assigned to orthogroups.</p>
  </li>
  <li>
    <p><strong>Statistics_PerSpecies.tsv</strong><br /> 
A tab separated text file that contains the same
information as the <code class="language-plaintext highlighter-rouge">Statistics_Overall.csv</code> file but for each individual species.</p>
  </li>
</ul>

<h3 id="gene_duplication_events">Gene_Duplication_Events</h3>

<ul>
  <li>
    <p><strong>Duplications.tsv</strong><br />
A tab separated text file that lists all the gene duplication events identified by examining each node of each orthogroup gene tree. The columns are:</p>

    <ol type="a">
  <li><strong><code>Orthogroup</code></strong></li>
  <li><strong><code>Species Tree node</code></strong>: see <code>SpeciesTree_rooted_node_labels.txt</code></li>
  <li><strong><code>Gene tree node</code></strong>: see orthogroup tree in <code>Resolved_Gene_Trees</code> directory</li>
  <li><strong><code>Support</code></strong>: proportion of genes for which this node is supported</li>
  <li><strong><code>Type</code></strong>:
  <ol type="i">
    <li><strong><code>Terminal</code></strong>: duplication on a terminal branch of the species tree</li>
    <li><strong><code>Non-Terminal</code></strong>: duplication on an internal branch of the species tree &amp; therefore shared by more than one species</li>
    <li><strong><code>Non-Terminal: STRIDE</code></strong>: Non-Terminal duplication that also passes the very stringent STRIDE checks for what the topology of the gene tree should be post-duplication.</li>
  </ol>
  </li>
  <li><strong><code>Genes 1</code></strong>: the list of genes descended from one of the copies of the duplicate gene</li>
  <li><strong><code>Genes 2</code></strong>: the list of genes descended from the other copy of the duplicate gene.</li>
</ol>
  </li>
  <li>
    <p><strong>SpeciesTree_Gene_Duplications_0.5_Support.txt</strong><br /> 
It provides a summation of the
above duplications over the branches of the species tree. The numbers after each
node or species name are the number of gene duplication events with at least 50%
support that occurred on the branch leading to the node/species. The branch lengths
are as given in Species_Tree/SpeciesTree_rooted.txt. It is a text file, in newick
format.</p>
  </li>
</ul>

<h3 id="orthogroup-sequences">Orthogroup Sequences</h3>

<p>A FASTA file for each orthogroup giving the amino acid sequences for each gene in
the orthogroup.</p>

<h3 id="orthogroups-directory">Orthogroups Directory</h3>

<ul>
  <li>
    <p><strong>Orthogroups.tsv</strong><br /> 
A tab separated text file that contains the genes in each orthogroup, with columns for each species.</p>
  </li>
  <li>
    <p><strong>Orthogroups.txt</strong><br /> 
A text file that contains the genes in each orthogroup (one orthogroup
per line).</p>
  </li>
  <li>
    <p><strong>Orthogroups_UnassignedGenes.tsv</strong><br /> 
A tab separated text file that contains all of
the genes that were not assigned to any orthogroup.</p>
  </li>
  <li>
    <p><strong>Orthogroups.GeneCount.tsv</strong><br /> 
A tab separated text file that contains counts of the number of genes for each species in each orthogroup.</p>
  </li>
  <li>
    <p><strong>Orthogroups_SingleCopyOrthologues.txt</strong><br /> 
A list of orthogroups that contain
exactly one gene per species i.e. they contain one-to-one orthologues.</p>
  </li>
</ul>

<h3 id="orthologues-directory">Orthologues Directory</h3>

<p>One sub-directory for each species that in turn contains a file for each pairwise species
comparison, listing the orthologs between that species pair.</p>

<p>Orthologues can be one-to-one, one-to-many or many-to-many depending on the gene
duplication events since the orthologs diverged.</p>

<p>Each row in a file contains the gene(s) in one species that are orthologues of the gene(s) in
the other species and each row is cross-referenced to the orthogroup that contains those
genes.</p>

<h3 id="phylogenetic-hierarchical-orthogroups-directory">Phylogenetic Hierarchical Orthogroups Directory</h3>

<p>OrthoFinder infers HOGs, orthogroups at each hierarchical level (i.e. at each node in the
species tree) by analysing the rooted gene trees.</p>

<ul>
  <li>
    <p><strong>N1.tsv</strong><br />
It is a tab separated text file. Each row contains the genes belonging to a single
orthogroup. The genes from each orthogroup are organized into columns, one per
species. Additional columns give the HOG (Hierarchical Orthogroup) ID and the node
in the gene tree from which the HOG was determined</p>
  </li>
  <li>
    <p><strong>N2.tsv, N3.tsv</strong>, …<br />
Orthogroups inferred from the gene trees corresponding to the clades of species in the species tree N2, N3, etc.</p>
  </li>
</ul>

<h3 id="phylogenetically_misplaced_genes">Phylogenetically_Misplaced_Genes</h3>

<p>Genes in “Phylogenetically_Misplaced_Genes/” are those that appear to be out of place in
the gene tree, and would otherwise negatively affect orthology analysis if not identified.</p>

<h3 id="putative_xenologs">Putative_Xenologs</h3>

<p>Xenologs are sets of genes descended from a common ancestor, but where there has been
horizontal transfer on the evolutionary path to the gene copies in extant species, rather than
just speciation and duplication. OrthoFinder tries to identify xenologs, but we call them
<code class="language-plaintext highlighter-rouge">putative</code>, since many arise from contamination during sequencing. Each species has a file
in this folder, listing genes and their putative xenologs from all other species</p>

<h3 id="resolved-gene-trees-directory">Resolved Gene Trees Directory</h3>

<p>A rooted phylogenetic tree inferred for each orthogroup with 4 or more sequences and
resolved using the OrthoFinder hybrid species-overlap/duplication-loss coalescent
model.</p>

<h3 id="single-copy-orthologue-sequences">Single Copy Orthologue Sequences</h3>

<p>The same files as the “Orthogroup Sequences” directory but restricted to only
those orthogroups which contain exactly one gene per species.</p>

<h3 id="species-tree-directory">Species Tree Directory</h3>

<ul>
  <li>
    <p><strong>SpeciesTree_rooted.txt</strong><br /> 
A STAG species tree inferred from all orthogroups, containing STAG support values at internal nodes and rooted using STRIDE.</p>
  </li>
  <li>
    <p><strong>SpeciesTree_rooted_node_labels.txt</strong><br />
The same tree as above but with the
nodes given labels (instead of support values) to allow other results files to crossreference
branches/nodes in the species tree (e.g. location of gene duplication
events).</p>
  </li>
</ul>

<h3 id="workingdirectory">WorkingDirectory</h3>

<p>This contains all the files necessary for orthofinder to run</p>

</section>
            </div> -->
        </li><li>
            <!-- <h2>
                <a class="post-link" href="">
                    <a href="">👉 Understanding Orthology</a>
                </a>
            </h2> -->

            <h2>
                <a class="post-link" href="/OrthoFinder/tutorials/understand-orthology/">
                    <span class="post-number">6.</span>
                    Understanding Orthology
                </a>
            </h2>

            <!-- <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i>
                

 4 minute read   
            </p> -->
            <p class="read-time">
                <i class="fa fa-clock-o" aria-hidden="true"></i> 
                

 4 minute read    | <!-- 




<span class="meta-info">
    🕒 Updated 20231 days ago

    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 20231 days ago
</span> -->







<span class="meta-info">
    <i class="fa fa-clock-o" aria-hidden="true"></i>
    Updated 18 days ago
</span>
            </p>

            <!-- <p class="post-meta-bar">
                <a href="/OrthoFinder/assets/docs/understand-orthology.pdf" target="_blank">
                    📄 Read in PDF
                </a> | 
                <a href="https://github.com/OrthoFinder/OrthoFinder/releases/download/v3.1.0/understand-orthology.pdf" target="_blank">
                    📄 Download PDF
                </a> | 
                <script>
    document.addEventListener("DOMContentLoaded", () => {
        fetch("https://api.github.com/repos/OrthoFinder/OrthoFinder/releases")
            .then(response => response.json())
            .then(data => {
                const assets = data.flatMap(release => release.assets);
                document.querySelectorAll(".download-count").forEach(el => {
                    const assetName = el.getAttribute("data-asset");
                    const match = assets.find(a => a.name === assetName);
                    if (match && match.download_count !== undefined) {
                        el.innerHTML = `<i class="fa fa-download" aria-hidden="true"></i> ${match.download_count} downloads`;
                    } else {
                        el.textContent = "(not found)";
                    }
                });
            })
            .catch(error => {
                console.error("GitHub API fetch error:", error);
                document.querySelectorAll(".download-count").forEach(el => {
                    el.textContent = "(error)";
                });
            });
    });
</script>
            </p> -->


            <!-- <ul>
                
                <li><a href="/OrthoFinder/tutorials/beginner-tutorial/">Beginner Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-tutorial/">Advanced Tutorial</a></li>
                
                <li><a href="/OrthoFinder/tutorials/advanced-usage/">Advanced Usage</a></li>
                
                <li><a href="/OrthoFinder/tutorials/gettting-input-data/">Getting Input Data</a></li>
                
                <li><a href="/OrthoFinder/tutorials/guide-to-results/">Guide to Results</a></li>
                
                <li><a href="/OrthoFinder/tutorials/understand-orthology/">Understanding Orthology</a></li>
                
            </ul> -->


            <!-- <span class="post-meta">May 1, 2025</span> -->

            <!-- <div class="container">
                <section id="tutorials"><h3 id="orthogroups-orthologs-and-paralogs">Orthogroups, Orthologs, and Paralogs</h3>

<ul class="tight-bullets">
  <li>
    <div class="bullet-header">
      <span class="bullet">•</span>
      <span class="bullet-title">Orthogroup</span>
    </div>
    <div class="bullet-text">
    The group of genes descended from a single gene in the last common ancestor
    of a group of species.
    </div>
  </li>
  <li>
    <div class="bullet-header">
      <span class="bullet">•</span>
      <span class="bullet-title">Orthologs</span>
    </div>
    <div class="bullet-text">
    Pairs of genes that descended from a single gene in the last common ancestor
    of two species.
    </div>
  </li>
  <li>
    <div class="bullet-header">
      <span class="bullet">•</span>
      <span class="bullet-title">Paralogs</span>
    </div>
    <div class="bullet-text">
    Pairs of genes descended from a gene duplication event.
    </div>
  </li>
</ul>

<p>Orthologs can be thought of as <code class="language-plaintext highlighter-rouge">equivalent genes</code> between two species, as they descended
from a single gene in the last common ancestor of that species. For example, the last
common ancestor of humans and mice is a small mammal which lived alongside the
dinosaurs. Individual genes present in that ancestor still exist in some form through their
descendants in both humans and mice, and those genes are orthologs.</p>

<p>Orthologs describe relationships between pairs of species, but we can extend this idea to
larger groups of species. Humans, mice, and chickens share a common ancestor from a few
hundred million years ago, before the dinosaurs had even emerged. We can describe a
group of genes across all three species that were descended from a single gene in this
ancestor - these genes form an orthogroup.</p>

<p>Look at the figure below, which shows data for three species: human, mouse and chicken.</p>

<p align="center" class="figure-wrapper">
  <img src="/OrthoFinder/assets/images/Understanding_image1.png" alt="understanding orthology" width="800" />
</p>
<p align="center" class="figure-caption"><strong>Figure 1: Orthologues, Orthogroups &amp; Paralogues</strong></p>

<p>The tree in Figure 1A shows the evolutionary history of a gene. First, there was a speciation
event where the chicken lineage diverged from the human-mouse ancestor. In the
human-mouse ancestor, there was a gene duplication event at X producing two copies of the
gene in that ancestor, Y &amp; Z. When human and mouse diverged they each inherited gene Y
(becoming HuA &amp; MoA) and gene Z (HuB &amp; MoB). In general, we can identify a gene
duplication event because it creates two copies of a gene in a species (e.g. HuA &amp; HuB).</p>

<p>The mouse gene MoB is closer related to a human gene than it is to the other mouse gene
MoA. This is because the gene duplication occurred in the ancestor, so each mouse gene is
more closely related to its human ortholog (e.g. HuA and MoA), as they were both
descended from a single gene. By contrast, HuA and MoB diverged at the gene
duplication event. They aren’t descended from a single gene in the common ancestor of
human and mice, so aren’t orthologs. Instead, they are paralogs (Figure 1C). Paralogs
are more distantly related, they diverged at a gene duplication event in a common
ancestor. Such a gene duplication event must have occurred further back in time than
when the species diverged and so paralogs between a pair of species are always less
closely related than orthologs between that pair of species. Paralogs are also possible
within a species (e.g. HuA &amp; HuB).</p>

<p>The chicken gene diverged from the other genes when the lineage leading to chicken
split from the lineage leading to human and mouse. Therefore, the chicken gene ChC is
an ortholog of HuA &amp; HuB in human and an ortholog of MoA &amp; MoB in mouse.
Depending on what happend after the genes diverged, orthologs can be in one-to-one
relationships (HuA - MoA), many-to-one (HuA &amp; HuB - ChC), or many-to-many (no
examples in this tree, but would occur if there were a duplication in chicken). All of these
relationships are identified by OrthoFinder.</p>

<h3 id="why-orthogroups">Why Orthogroups?</h3>

<p>There are several reasons why Orthogroups are the relevant way of analysing orthology
relationships between species:</p>

<ul class="tight-bullets">
  <li>
    <div class="bullet-header">
      <span class="bullet">•</span>
      <span class="bullet-title">Orthogroups allow you to analyse all of your data</span>
    </div>
    <div class="bullet-text">
All of the genes in an orthogroup are descended from a single ancestral gene. Thus, all
the genes in an orthogroup started out with the same sequence and function. As gene
duplication and loss occur frequently in evolution, one-to-one orthologs are rare and
limitation of analyses to on-to-one orthologs limits an analysis to a small fraction of the
available data. By analysing orthogroups you can analyse all of your data.
    </div>
  </li>
  <li>
    <div class="bullet-header">
      <span class="bullet">•</span>
      <span class="bullet-title">Orthogroups allow you to define the unit of comparison</span>
    </div>
    <div class="bullet-text">
It is important to note that with orthogroups you choose where to define the limits of the
unit of comparison. For example, if you just chose to analyse human and mouse in the
above figure then you would have two orthogroups.
    </div>
  </li>
  <li>
    <div class="bullet-header">
      <span class="bullet">•</span>
      <span class="bullet-title">Orthogroups are the only way to identify orthologs</span>
    </div>
    <div class="bullet-text">
Orthology is defined by phylogeny. It is not definable by amino acid content, codon bias,
GC content or other measures of sequence similarity. Methods that use such scores to
define orthologs in the absence of phylogeny can only provide guesses. The only way to
be sure that the orthology assignment is correct is by conducting a phylogenetic
reconstruction of all genes descended from a single gene the last common ancestor of
the species under consideration. This set of genes is an orthogroup. Thus, the only way
to define orthology is by analysing orthogroups.
    </div>
  </li>
</ul>

<p>For a comprehensive overview of orthology and the OrthoFinder approach, you can
watch David Emms’ conference talk on <a href="https://www.youtube.com/watch?v=L6eXJAE5J7g">YouTube</a>, from the 2020 symposium on Phylogenomics and
Comparative genomics.</p>

</section>
            </div> -->
        </li></ul>


    </div>
        
      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/OrthoFinder/"></data>

  <div class="wrapper">

    <div class="footer-col-wrapper">
      <div class="footer-col">

        <!-- <p class="feed-subscribe">
          <a href="http://localhost:4000/OrthoFinder/feed.xml">
            <svg class="svg-icon orange">
              <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#rss"></use>
            </svg><span>Subscribe</span>
          </a>
        </p> -->

        <!--<ul class="social-media-list"><li>
  <a rel="me" href="" target="_blank" title="GitHub">
    <svg class="svg-icon grey">
      <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
    </svg>
  </a>
</li>


</ul>
-->
         
        <div class="social-icons">
          
          <a href="https://github.com/OrthoFinder/OrthoFinder" target="_blank" title="GitHub">
            <svg class="svg-icon grey">
              <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
            </svg>
            <span>GitHub</span>
          </a>
          
          <!--  -->
        </div>


          <!--
          <ul class="contact-list">

            Link to the team section instead of listing every author
            <li>
              <a href="/OrthoFinder/about/#team">
                Meet&nbsp;the&nbsp;Team
              </a>
            </li>

          </ul> -->

          <ul class="contact-list">
            <li><a href="/OrthoFinder/about/#team">Meet the Team</a></li>
          </ul>


      </div>
      
      <div class="footer-col">
        <!-- <p>This is a software developed by the [Steven Kelly Lab](http://www.stevekellylab.com/).</p> -->
        <p><p>This is a software developed by the <a href="http://www.stevekellylab.com/">Steven Kelly Lab</a>.</p>
</p>
      </div>
    </div>

    <!-- <div class="social-links"><ul class="social-media-list"><li>
  <a rel="me" href="" target="_blank" title="GitHub">
    <svg class="svg-icon grey">
      <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
    </svg>
  </a>
</li>


</ul>
</div> -->

  </div>


<script>
  document.addEventListener('DOMContentLoaded', function () {
    // 1) Find all <pre><code> blocks
    const codeBlocks = document.querySelectorAll('pre > code');

    codeBlocks.forEach((codeBlock, index) => {
      // The parent <pre> element
      const pre = codeBlock.parentNode;

      // 2) Create a container DIV and wrap <pre> inside it
      const container = document.createElement('div');
      container.className = 'code-container';
      pre.parentNode.insertBefore(container, pre);
      container.appendChild(pre);

      // 3) Create the copy button
      const button = document.createElement('button');
      button.className = 'copy-button';
      button.type = 'button';

      // 4) Insert an inline SVG icon + "Copy" text
      //    We'll build the SVG as a string, then inject it into button.innerHTML
      const svgIcon = `
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="16"
          height="16"
          fill="currentColor"
          class="bi bi-files"
          viewBox="0 0 16 16"
        >
            <path d="M13 0H5a1 1 0 0 0-1 1v1H3a2 2 0 0 0-2
                  2v9.5A1.5 1.5 0 0 0 2.5 15H9v-1H2.5a.5.5
                  0 0 1-.5-.5V4a1 1 0 0 1 1-1h1v7.5A1.5
                  1.5 0 0 0 5.5 12H11v-1H5.5a.5.5 0 0
                  1-.5-.5V1a.5.5 0 0 1 .5-.5H5v3.5A1.5
                  1.5 0 0 0 6.5 5H13v6h1V1a1 1 0 0 0-1-1z"/>
        </svg>
    `;
      

      
      button.innerHTML = svgIcon + 'Copy';

      // 5) Assign an ID to <code> if it doesn't have one
      if (!codeBlock.id) {
        codeBlock.id = `codeblock-${index}`;
      }

      // 6) Link button to the code block by ID
      button.setAttribute('data-clipboard-target', '#' + codeBlock.id);

      // 7) Insert the button into the container
      container.insertBefore(button, pre);
    });

    // 8) Initialize Clipboard.js for all .copy-button elements
    const clipboard = new ClipboardJS('.copy-button');
    clipboard.on('success', function (e) {
      e.clearSelection();
      // Briefly change button text to indicate success
      e.trigger.innerText = 'Copied!';
      setTimeout(() => {
        e.trigger.innerHTML = svgIcon + 'Copy'; // restore icon + text
      }, 1500);
    });
    clipboard.on('error', function (e) {
      console.error('Failed to copy', e);
    });
  });
</script>


  <!-- <script>
    document.addEventListener('DOMContentLoaded', function () {
      // 1. Find all <pre><code> blocks
      const codeBlocks = document.querySelectorAll('pre > code');

      codeBlocks.forEach((codeBlock) => {
        const pre = codeBlock.parentNode;

        // 2. Create a container div to hold the code + button
        const container = document.createElement('div');
        container.className = 'code-container';

        // Insert the container before <pre>, then move <pre> inside it
        pre.parentNode.insertBefore(container, pre);
        container.appendChild(pre);

        // 3. Create a "Copy" button
        const button = document.createElement('button');
        button.className = 'copy-button';
        button.type = 'button';
        button.innerText = 'Copy';

        // If the code block doesn’t have an ID, assign one so we can target it
        let codeId = codeBlock.id;
        if (!codeId) {
          codeId = 'codeblock-' + Math.floor(Math.random() * 1000000);
          codeBlock.setAttribute('id', codeId);
        }

        // 4. Point the button at the code block
        button.setAttribute('data-clipboard-target', '#' + codeId);

        // 5. Insert button into the container
        container.insertBefore(button, pre);
      });

      // 6. Initialize Clipboard.js
      const clipboard = new ClipboardJS('.copy-button');
      clipboard.on('success', function (e) {
        e.clearSelection();
        // Change button text to confirm copy
        e.trigger.textContent = 'Copied!';
        setTimeout(() => {
          e.trigger.textContent = 'Copy';
        }, 2000);
      });
      clipboard.on('error', function (e) {
        console.error('Failed to copy', e);
      });
    });
  </script> -->


</footer></body>

</html>

