<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Getting Input Data | OrthoFinder</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="Getting Input Data" />
<meta name="author" content="David Emms, Jonathan Holmes, Laurence Belcher, Yi Liu" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Downloading input files for OrthoFinder Let’s imagine we want to perform a phylogenomic analysis across a set of species: Dingo (Canis lupus dingo), Great spotted kiwi (Apteryx haastii), Kakapo (Strigops habroptila), Platypus (Ornithorhynchus anatinus), Tammar wallaby (Notamacropus eugenii), and Common wombat (Vombatus ursinus). OrthoFinder requires as input the amino acid sequences for all the protein coding genes in your species of interest. In this section of the tutorial we will discuss how to get these files for the species that you want to analyse. We will cover three major websites for getting this data, Ensembl, NCBI, and Phytozome. We recommend Ensembl if you want to analyse eukaryotic species, especially vertebrates and model organisms. We recommend NCBI if you want to analyse prokaryotic species. We recommend Phytozome if you want to analyse plants. A key consideration when getting input data for OrthoFinder is gene transcripts. When you download the amino acid sequences for a genome, you might have multiple sequences for the same gene. This is because a single gene can produce multiple transcripts, which each might produce a different protein. You can see an example of transcripts in this table of the Actin gene on ensembl. If we ran OrthoFinder on these raw files it would take much longer than necessary, and could lower the accuracy. We therefore want to extract just the longest transcript variant for each gene. OrthoFinder provides scripts to do this, which we will learn how to use later. Getting data from Ensembl First, we go to the ensembl webpage with the list of species https://www.ensembl.org/info/about/species.html You can also use the new beta website, which is updated more regularly https://beta.ensembl.org/species-selector We can then search for our first species (Canis lupus dingo), and click on the link This will take us to the ensembl page for that species input_img1 On the right hand side under the heading ‘Gene annotation’ we can press the link to ‘Download FASTA’ files for the genome. This will take us to a webpage with some folders. input_img2 We need to click on the ‘pep/’ folder, and then click on the file that ends in .pep.all.fa.gz input_img3 We can then repeat this step for our other species, and place the files that are downloaded into a folder on our computer. I have named my folder ‘Proteomes’. The files are stored as .gz compressed files to save space, but we now need to expand the files so that we can access the sequences. You can either double-click on them all, or use the command line gunzip *.gz We’ll use a script provided with OrthoFinder to extract just the longest transcript variant per gene and run OrthoFinder on these files: You can find the script here To run the script, first place it in the Proteomes folder. Then, open the command line and navigate to the Proteomes folder (using cd). You can then use the following command to run the script for f in *fa ; do python primary_transcript.py $f ; done The script will generate a new folder called ‘primary_transcripts’, which contains our files. Shortening the filenames is a good idea as it keeps the results tidy as the filenames are used to refer to the species, e.g. I shortened it to Canis_lupus_dingo.fa. Our data is now ready for OrthoFinder. You can go to the ‘Beginner Tutorial’ to see how to run OrthoFinder, or check out the other guides on getting input data from other sources Getting data from NCBI We start by going to https://www.ncbi.nlm.nih.gov/ and searching for our first species input_img4 We want to find the genome of this species, so scroll down to the ‘Genomes’ section, and click on ‘Assembly / Genome’ input_img5 There might be several genomes listed on this page. We are going to click on the one with the green tick, which shows us the reference genome. input_img6 We can then click the blue ‘Download’ button, and click the boxes to select the curated ‘RefSeq only’ annotation, and that we want the GFF and protein FASTA files. We can also give our download a useful name (such as the name of the species). input_img7 We can then click the ‘Download’ button, which downloads a zip file. Repeat this step for all of the species that you want to use. For some species in our example list, like the Kiwi, there is no RefSeq genome with protein fasta available. For these species, it is recommended to use Ensembl. Now, we need to deal with the issue of multiple transcripts per gene. For our first species, the Dingo, there are about 20,000 genes. However, if we look at the ‘protein.faa’ file that we have downloaded, it has about 75,000 sequences. We can use the script ncbi_primary_transcript.py to extract the longest transcript per gene. Place the script in the folder that has the zip folders, and run the following line of code in the command line python ncbi_primary_transcripts.py This will then make a folder named ‘primary_transcripts’, which is ready to run Orthofinder on. If you want to download data in bulk, you can also use the NCBI Datasets tool, which can run on the command line. For more info, see here Getting data from Phytozome First, go to the Phytozome data portal website https://data.jgi.doe.gov/ Here, you can search for a species, such as Oryza sativa input_img8 You can then find the genome that want to download, and click on it The file that we want ends in ‘primaryTranscriptOnly.fa.gz’. Luckily for us, Phytozome already deals with the problem of multiple transcripts. We can select that file, and click ‘Add to Cart’. input_img9 We can then repeat this step for all of the species that we want to download. When we are ready, we then click the shopping cart icon, and press ‘Download’ on the next page. You will have to sign-in or register to download data. These files can be placed in a folder, where they are now ready to run OrthoFinder." />
<meta property="og:description" content="Downloading input files for OrthoFinder Let’s imagine we want to perform a phylogenomic analysis across a set of species: Dingo (Canis lupus dingo), Great spotted kiwi (Apteryx haastii), Kakapo (Strigops habroptila), Platypus (Ornithorhynchus anatinus), Tammar wallaby (Notamacropus eugenii), and Common wombat (Vombatus ursinus). OrthoFinder requires as input the amino acid sequences for all the protein coding genes in your species of interest. In this section of the tutorial we will discuss how to get these files for the species that you want to analyse. We will cover three major websites for getting this data, Ensembl, NCBI, and Phytozome. We recommend Ensembl if you want to analyse eukaryotic species, especially vertebrates and model organisms. We recommend NCBI if you want to analyse prokaryotic species. We recommend Phytozome if you want to analyse plants. A key consideration when getting input data for OrthoFinder is gene transcripts. When you download the amino acid sequences for a genome, you might have multiple sequences for the same gene. This is because a single gene can produce multiple transcripts, which each might produce a different protein. You can see an example of transcripts in this table of the Actin gene on ensembl. If we ran OrthoFinder on these raw files it would take much longer than necessary, and could lower the accuracy. We therefore want to extract just the longest transcript variant for each gene. OrthoFinder provides scripts to do this, which we will learn how to use later. Getting data from Ensembl First, we go to the ensembl webpage with the list of species https://www.ensembl.org/info/about/species.html You can also use the new beta website, which is updated more regularly https://beta.ensembl.org/species-selector We can then search for our first species (Canis lupus dingo), and click on the link This will take us to the ensembl page for that species input_img1 On the right hand side under the heading ‘Gene annotation’ we can press the link to ‘Download FASTA’ files for the genome. This will take us to a webpage with some folders. input_img2 We need to click on the ‘pep/’ folder, and then click on the file that ends in .pep.all.fa.gz input_img3 We can then repeat this step for our other species, and place the files that are downloaded into a folder on our computer. I have named my folder ‘Proteomes’. The files are stored as .gz compressed files to save space, but we now need to expand the files so that we can access the sequences. You can either double-click on them all, or use the command line gunzip *.gz We’ll use a script provided with OrthoFinder to extract just the longest transcript variant per gene and run OrthoFinder on these files: You can find the script here To run the script, first place it in the Proteomes folder. Then, open the command line and navigate to the Proteomes folder (using cd). You can then use the following command to run the script for f in *fa ; do python primary_transcript.py $f ; done The script will generate a new folder called ‘primary_transcripts’, which contains our files. Shortening the filenames is a good idea as it keeps the results tidy as the filenames are used to refer to the species, e.g. I shortened it to Canis_lupus_dingo.fa. Our data is now ready for OrthoFinder. You can go to the ‘Beginner Tutorial’ to see how to run OrthoFinder, or check out the other guides on getting input data from other sources Getting data from NCBI We start by going to https://www.ncbi.nlm.nih.gov/ and searching for our first species input_img4 We want to find the genome of this species, so scroll down to the ‘Genomes’ section, and click on ‘Assembly / Genome’ input_img5 There might be several genomes listed on this page. We are going to click on the one with the green tick, which shows us the reference genome. input_img6 We can then click the blue ‘Download’ button, and click the boxes to select the curated ‘RefSeq only’ annotation, and that we want the GFF and protein FASTA files. We can also give our download a useful name (such as the name of the species). input_img7 We can then click the ‘Download’ button, which downloads a zip file. Repeat this step for all of the species that you want to use. For some species in our example list, like the Kiwi, there is no RefSeq genome with protein fasta available. For these species, it is recommended to use Ensembl. Now, we need to deal with the issue of multiple transcripts per gene. For our first species, the Dingo, there are about 20,000 genes. However, if we look at the ‘protein.faa’ file that we have downloaded, it has about 75,000 sequences. We can use the script ncbi_primary_transcript.py to extract the longest transcript per gene. Place the script in the folder that has the zip folders, and run the following line of code in the command line python ncbi_primary_transcripts.py This will then make a folder named ‘primary_transcripts’, which is ready to run Orthofinder on. If you want to download data in bulk, you can also use the NCBI Datasets tool, which can run on the command line. For more info, see here Getting data from Phytozome First, go to the Phytozome data portal website https://data.jgi.doe.gov/ Here, you can search for a species, such as Oryza sativa input_img8 You can then find the genome that want to download, and click on it The file that we want ends in ‘primaryTranscriptOnly.fa.gz’. Luckily for us, Phytozome already deals with the problem of multiple transcripts. We can select that file, and click ‘Add to Cart’. input_img9 We can then repeat this step for all of the species that we want to download. When we are ready, we then click the shopping cart icon, and press ‘Download’ on the next page. You will have to sign-in or register to download data. These files can be placed in a folder, where they are now ready to run OrthoFinder." />
<link rel="canonical" href="http://localhost:4000/OrthoFinder/tutorials/gettting-input-data/" />
<meta property="og:url" content="http://localhost:4000/OrthoFinder/tutorials/gettting-input-data/" />
<meta property="og:site_name" content="OrthoFinder" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2025-05-01T10:04:16+01:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Getting Input Data" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","author":{"@type":"Person","name":"David Emms, Jonathan Holmes, Laurence Belcher, Yi Liu"},"dateModified":"2025-05-01T10:04:16+01:00","datePublished":"2025-05-01T10:04:16+01:00","description":"Downloading input files for OrthoFinder Let’s imagine we want to perform a phylogenomic analysis across a set of species: Dingo (Canis lupus dingo), Great spotted kiwi (Apteryx haastii), Kakapo (Strigops habroptila), Platypus (Ornithorhynchus anatinus), Tammar wallaby (Notamacropus eugenii), and Common wombat (Vombatus ursinus). OrthoFinder requires as input the amino acid sequences for all the protein coding genes in your species of interest. In this section of the tutorial we will discuss how to get these files for the species that you want to analyse. We will cover three major websites for getting this data, Ensembl, NCBI, and Phytozome. We recommend Ensembl if you want to analyse eukaryotic species, especially vertebrates and model organisms. We recommend NCBI if you want to analyse prokaryotic species. We recommend Phytozome if you want to analyse plants. A key consideration when getting input data for OrthoFinder is gene transcripts. When you download the amino acid sequences for a genome, you might have multiple sequences for the same gene. This is because a single gene can produce multiple transcripts, which each might produce a different protein. You can see an example of transcripts in this table of the Actin gene on ensembl. If we ran OrthoFinder on these raw files it would take much longer than necessary, and could lower the accuracy. We therefore want to extract just the longest transcript variant for each gene. OrthoFinder provides scripts to do this, which we will learn how to use later. Getting data from Ensembl First, we go to the ensembl webpage with the list of species https://www.ensembl.org/info/about/species.html You can also use the new beta website, which is updated more regularly https://beta.ensembl.org/species-selector We can then search for our first species (Canis lupus dingo), and click on the link This will take us to the ensembl page for that species input_img1 On the right hand side under the heading ‘Gene annotation’ we can press the link to ‘Download FASTA’ files for the genome. This will take us to a webpage with some folders. input_img2 We need to click on the ‘pep/’ folder, and then click on the file that ends in .pep.all.fa.gz input_img3 We can then repeat this step for our other species, and place the files that are downloaded into a folder on our computer. I have named my folder ‘Proteomes’. The files are stored as .gz compressed files to save space, but we now need to expand the files so that we can access the sequences. You can either double-click on them all, or use the command line gunzip *.gz We’ll use a script provided with OrthoFinder to extract just the longest transcript variant per gene and run OrthoFinder on these files: You can find the script here To run the script, first place it in the Proteomes folder. Then, open the command line and navigate to the Proteomes folder (using cd). You can then use the following command to run the script for f in *fa ; do python primary_transcript.py $f ; done The script will generate a new folder called ‘primary_transcripts’, which contains our files. Shortening the filenames is a good idea as it keeps the results tidy as the filenames are used to refer to the species, e.g. I shortened it to Canis_lupus_dingo.fa. Our data is now ready for OrthoFinder. You can go to the ‘Beginner Tutorial’ to see how to run OrthoFinder, or check out the other guides on getting input data from other sources Getting data from NCBI We start by going to https://www.ncbi.nlm.nih.gov/ and searching for our first species input_img4 We want to find the genome of this species, so scroll down to the ‘Genomes’ section, and click on ‘Assembly / Genome’ input_img5 There might be several genomes listed on this page. We are going to click on the one with the green tick, which shows us the reference genome. input_img6 We can then click the blue ‘Download’ button, and click the boxes to select the curated ‘RefSeq only’ annotation, and that we want the GFF and protein FASTA files. We can also give our download a useful name (such as the name of the species). input_img7 We can then click the ‘Download’ button, which downloads a zip file. Repeat this step for all of the species that you want to use. For some species in our example list, like the Kiwi, there is no RefSeq genome with protein fasta available. For these species, it is recommended to use Ensembl. Now, we need to deal with the issue of multiple transcripts per gene. For our first species, the Dingo, there are about 20,000 genes. However, if we look at the ‘protein.faa’ file that we have downloaded, it has about 75,000 sequences. We can use the script ncbi_primary_transcript.py to extract the longest transcript per gene. Place the script in the folder that has the zip folders, and run the following line of code in the command line python ncbi_primary_transcripts.py This will then make a folder named ‘primary_transcripts’, which is ready to run Orthofinder on. If you want to download data in bulk, you can also use the NCBI Datasets tool, which can run on the command line. For more info, see here Getting data from Phytozome First, go to the Phytozome data portal website https://data.jgi.doe.gov/ Here, you can search for a species, such as Oryza sativa input_img8 You can then find the genome that want to download, and click on it The file that we want ends in ‘primaryTranscriptOnly.fa.gz’. Luckily for us, Phytozome already deals with the problem of multiple transcripts. We can select that file, and click ‘Add to Cart’. input_img9 We can then repeat this step for all of the species that we want to download. When we are ready, we then click the shopping cart icon, and press ‘Download’ on the next page. You will have to sign-in or register to download data. These files can be placed in a folder, where they are now ready to run OrthoFinder.","headline":"Getting Input Data","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/OrthoFinder/tutorials/gettting-input-data/"},"url":"http://localhost:4000/OrthoFinder/tutorials/gettting-input-data/"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/OrthoFinder/assets/css/style.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/OrthoFinder/feed.xml" title="OrthoFinder" /><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
</head>
<body><header class="site-header">

  <div class="wrapper"><!-- <a class="site-title" rel="author" href="/OrthoFinder/%20/">OrthoFinder</a> -->
    <a class="site-title" rel="author" href="/OrthoFinder/">OrthoFinder</a><nav class="site-nav">
      <input type="checkbox" id="nav-trigger" class="nav-trigger" />
      <label for="nav-trigger">
        <span class="menu-icon">
          <svg viewBox="0 0 18 15" width="18px" height="15px">
            <path
              d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z" />
          </svg>
        </span>
      </label>
      <div class="trigger">
        <!--<a class="page-link" href="/OrthoFinder/">Home</a><a class="page-link" href="/OrthoFinder/about/">About</a><a class="page-link" href="/OrthoFinder/download/">Download</a><a class="page-link" href="/OrthoFinder/installation/">Installation</a><a class="page-link" href="/OrthoFinder/tutorial/">Tutorial</a><a class="page-link" href="/OrthoFinder/faq/">FAQ</a>--><a class="page-link" href="/OrthoFinder/">Home</a><a class="page-link" href="/OrthoFinder/about/">About</a><a class="page-link" href="/OrthoFinder/download/">Download</a><a class="page-link" href="/OrthoFinder/installation/">Installation</a><a class="page-link" href="/OrthoFinder/tutorial/">Tutorial</a><a class="page-link" href="/OrthoFinder/faq/">FAQ</a>
        <a class="page-link" href="https://github.com/OrthoFinder/OrthoFinder" target="_blank" title="GitHub Profile">GitHub</a>
        
      </div>
    </nav></div>
  <script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js"></script>
</header><main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Getting Input Data</h1>
    <p class="post-meta"><!-- <time class="dt-published" datetime="2025-05-01T10:04:16+01:00" itemprop="datePublished">
        May 1, 2025
      </time> --><!-- <i class="fa fa-refresh" aria-hidden="true"></i> -->
      <time class="dt-modified" datetime="2025-05-06T00:00:00+01:00" itemprop="dateModified">
        Updated May 6, 2025
      </time>
      <meta itemprop="datePublished" content="2025-05-01T10:04:16+01:00"></p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <h2 id="downloading-input-files-for-orthofinder">Downloading input files for OrthoFinder</h2>

<p>Let’s imagine we want to perform a phylogenomic analysis across a set of species:
Dingo (<em>Canis lupus dingo</em>), Great spotted kiwi (<em>Apteryx haastii</em>), Kakapo (<em>Strigops
habroptila</em>), Platypus (<em>Ornithorhynchus anatinus</em>), Tammar wallaby (<em>Notamacropus
eugenii</em>), and Common wombat (<em>Vombatus ursinus</em>).</p>

<p>OrthoFinder requires as input the amino acid sequences for all the protein coding genes
in your species of interest. In this section of the tutorial we will discuss how to get these
files for the species that you want to analyse. We will cover three major websites for
getting this data, Ensembl, NCBI, and Phytozome.</p>

<p>We recommend <a href="#getting-data-from-ensembl"><strong>Ensembl</strong></a> if you want to analyse eukaryotic species, especially
vertebrates and model organisms.</p>

<p>We recommend <a href="#getting-data-from-ncbi"><strong>NCBI</strong></a> if you want to analyse prokaryotic species.</p>

<p>We recommend <a href="#getting-data-from-phytozome"><strong>Phytozome</strong></a> if you want to analyse plants.</p>

<p>A key consideration when getting input data for OrthoFinder is gene transcripts. When you download the amino acid sequences for a genome, you might have multiple
sequences for the same gene. This is because a single gene can produce multiple
transcripts, which each might produce a different protein. You can see an example of
transcripts in this table of the Actin gene on <a href="http://www.ensembl.org/Homo_sapiens/Gene/Splice?db=core;g=ENSG00000075624;r=7:5527151-5563784">ensembl</a>.</p>

<p>If we ran OrthoFinder on these raw files it would take much longer than necessary, and
could lower the accuracy. We therefore want to extract just the longest transcript variant for
each gene. OrthoFinder provides scripts to do this, which we will learn how to use later.</p>

<h2 id="getting-data-from-ensembl">Getting data from Ensembl</h2>

<p>First, we go to the ensembl webpage with the list of species
https://www.ensembl.org/info/about/species.html</p>

<p>You can also use the new beta website, which is updated more regularly
https://beta.ensembl.org/species-selector</p>

<p>We can then search for our first species (<em>Canis lupus dingo</em>), and click on the
link This will take us to the ensembl page for that species</p>

<p><em>input_img1</em></p>

<p>On the right hand side under the heading ‘Gene annotation’ we can press the link to
‘Download FASTA’ files for the genome.</p>

<p>This will take us to a webpage with some folders.</p>

<p><em>input_img2</em></p>

<p>We need to click on the ‘pep/’ folder, and then click on the file that ends in .pep.all.fa.gz</p>

<p><em>input_img3</em></p>

<p>We can then repeat this step for our other species, and place the files that are
downloaded into a folder on our computer. I have named my folder ‘Proteomes’.</p>

<p>The files are stored as .gz compressed files to save space, but we now need to expand
the files so that we can access the sequences. You can either double-click on them all,
or use the command line
<code class="language-plaintext highlighter-rouge">gunzip *.gz</code></p>

<p>We’ll use a script provided with OrthoFinder to extract just the longest transcript variant
per gene and run OrthoFinder on these files:</p>

<p>You can find the script <a href="https://github.com/OrthoFinder/OrthoFinder/blob/master/tools/primary_transcript.py">here</a></p>

<p>To run the script, first place it in the Proteomes folder. Then, open the command line
and navigate to the Proteomes folder (using cd). You can then use the following
command to run the script</p>

<p><code class="language-plaintext highlighter-rouge">for f in *fa ; do python primary_transcript.py $f ; done</code></p>

<p>The script will generate a new folder called ‘primary_transcripts’, which contains our files.
Shortening the filenames is a good idea as it keeps the results tidy as the filenames are
used to refer to the species, e.g. I shortened it to Canis_lupus_dingo.fa.</p>

<p>Our data is now ready for OrthoFinder. You can go to the ‘Beginner Tutorial’ to see how to run OrthoFinder,
or check out the other guides on getting input data from other sources</p>

<h2 id="getting-data-from-ncbi">Getting data from NCBI</h2>

<p>We start by going to https://www.ncbi.nlm.nih.gov/ and searching for our first species</p>

<p><em>input_img4</em></p>

<p>We want to find the genome of this species, so scroll down to the ‘Genomes’ section,
and click on ‘Assembly / Genome’</p>

<p><em>input_img5</em></p>

<p>There might be several genomes listed on this page. We are going to click on the one
with the green tick, which shows us the reference genome.</p>

<p><em>input_img6</em></p>

<p>We can then click the blue ‘Download’ button, and click the boxes to select the curated
‘RefSeq only’ annotation, and that we want the GFF and protein FASTA files. We can
also give our download a useful name (such as the name of the species).</p>

<p><em>input_img7</em></p>

<p>We can then click the ‘Download’ button, which downloads a zip file.</p>

<p>Repeat this step for all of the species that you want to use. For some species in our
example list, like the Kiwi, there is no RefSeq genome with protein fasta available. For
these species, it is recommended to use Ensembl.</p>

<p>Now, we need to deal with the issue of multiple transcripts per gene. For our first
species, the Dingo, there are about 20,000 genes. However, if we look at the
‘protein.faa’ file that we have downloaded, it has about 75,000 sequences.</p>

<p>We can use the script <code class="language-plaintext highlighter-rouge">ncbi_primary_transcript.py</code> to extract the longest
transcript per gene.</p>

<p>Place the script in the folder that has the zip folders, and run the following line of code in the
command line</p>

<p><code class="language-plaintext highlighter-rouge">python ncbi_primary_transcripts.py</code></p>

<p>This will then make a folder named ‘primary_transcripts’, which is ready to run
Orthofinder on.</p>

<p>If you want to download data in bulk, you can also use the NCBI Datasets tool, which
can run on the command line. For more info, see <a href="https://www.ncbi.nlm.nih.gov/datasets/">here</a></p>

<h2 id="getting-data-from-phytozome">Getting data from Phytozome</h2>

<p>First, go to the Phytozome data portal website https://data.jgi.doe.gov/ Here, you can search for a species, such
as Oryza sativa</p>

<p><em>input_img8</em></p>

<p>You can then find the genome that want to download, and click on it</p>

<p>The file that we want ends in ‘primaryTranscriptOnly.fa.gz’. Luckily for us, Phytozome
already deals with the problem of multiple transcripts. We can select that file, and click
‘Add to Cart’.</p>

<p><em>input_img9</em></p>

<p>We can then repeat this step for all of the species that we want to download. When we
are ready, we then click the shopping cart icon, and press ‘Download’ on the next page.
You will have to sign-in or register to download data.</p>

<p>These files can be placed in a folder, where they are now ready to run OrthoFinder.</p>


  </div><a class="u-url" href="/OrthoFinder/tutorials/gettting-input-data/" hidden></a>
</article>
<script>
    document.addEventListener("DOMContentLoaded", () => {
        fetch("https://api.github.com/repos/OrthoFinder/OrthoFinder/releases")
            .then(response => response.json())
            .then(data => {
                const assets = data.flatMap(release => release.assets);
                document.querySelectorAll(".download-count").forEach(el => {
                    const assetName = el.getAttribute("data-asset");
                    const match = assets.find(a => a.name === assetName);
                    if (match && match.download_count !== undefined) {
                        el.innerHTML = `<i class="fa fa-download" aria-hidden="true"></i> ${match.download_count} downloads`;
                    } else {
                        el.textContent = "(not found)";
                    }
                });
            })
            .catch(error => {
                console.error("GitHub API fetch error:", error);
                document.querySelectorAll(".download-count").forEach(el => {
                    el.textContent = "(error)";
                });
            });
    });
</script>
        
      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/OrthoFinder/"></data>

  <div class="wrapper">

    <div class="footer-col-wrapper">
      <div class="footer-col">

        <!-- <p class="feed-subscribe">
          <a href="http://localhost:4000/OrthoFinder/feed.xml">
            <svg class="svg-icon orange">
              <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#rss"></use>
            </svg><span>Subscribe</span>
          </a>
        </p> -->

        <!--<ul class="social-media-list"><li>
  <a rel="me" href="" target="_blank" title="GitHub">
    <svg class="svg-icon grey">
      <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
    </svg>
  </a>
</li>


</ul>
-->
         
        <div class="social-icons">
          
          <a href="https://github.com/OrthoFinder/OrthoFinder" target="_blank" title="GitHub">
            <svg class="svg-icon grey">
              <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
            </svg>
            <span>GitHub</span>
          </a>
          
          <!--  -->
        </div>


          <!--
          <ul class="contact-list">

            Link to the team section instead of listing every author
            <li>
              <a href="/OrthoFinder/about/#team">
                Meet&nbsp;the&nbsp;Team
              </a>
            </li>

          </ul> -->

          <ul class="contact-list">
            <li><a href="/OrthoFinder/about/#team">Meet the Team</a></li>
          </ul>


      </div>
      
      <div class="footer-col">
        <!-- <p>This is a software developed by the [Steven Kelly Lab](http://www.stevekellylab.com/).</p> -->
        <p><p>This is a software developed by the <a href="http://www.stevekellylab.com/">Steven Kelly Lab</a>.</p>
</p>
      </div>
    </div>

    <!-- <div class="social-links"><ul class="social-media-list"><li>
  <a rel="me" href="" target="_blank" title="GitHub">
    <svg class="svg-icon grey">
      <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
    </svg>
  </a>
</li>


</ul>
</div> -->

  </div>


<script>
  document.addEventListener('DOMContentLoaded', function () {
    // 1) Find all <pre><code> blocks
    const codeBlocks = document.querySelectorAll('pre > code');

    codeBlocks.forEach((codeBlock, index) => {
      // The parent <pre> element
      const pre = codeBlock.parentNode;

      // 2) Create a container DIV and wrap <pre> inside it
      const container = document.createElement('div');
      container.className = 'code-container';
      pre.parentNode.insertBefore(container, pre);
      container.appendChild(pre);

      // 3) Create the copy button
      const button = document.createElement('button');
      button.className = 'copy-button';
      button.type = 'button';

      // 4) Insert an inline SVG icon + "Copy" text
      //    We'll build the SVG as a string, then inject it into button.innerHTML
      const svgIcon = `
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="16"
          height="16"
          fill="currentColor"
          class="bi bi-files"
          viewBox="0 0 16 16"
        >
            <path d="M13 0H5a1 1 0 0 0-1 1v1H3a2 2 0 0 0-2
                  2v9.5A1.5 1.5 0 0 0 2.5 15H9v-1H2.5a.5.5
                  0 0 1-.5-.5V4a1 1 0 0 1 1-1h1v7.5A1.5
                  1.5 0 0 0 5.5 12H11v-1H5.5a.5.5 0 0
                  1-.5-.5V1a.5.5 0 0 1 .5-.5H5v3.5A1.5
                  1.5 0 0 0 6.5 5H13v6h1V1a1 1 0 0 0-1-1z"/>
        </svg>
    `;
      

      
      button.innerHTML = svgIcon + 'Copy';

      // 5) Assign an ID to <code> if it doesn't have one
      if (!codeBlock.id) {
        codeBlock.id = `codeblock-${index}`;
      }

      // 6) Link button to the code block by ID
      button.setAttribute('data-clipboard-target', '#' + codeBlock.id);

      // 7) Insert the button into the container
      container.insertBefore(button, pre);
    });

    // 8) Initialize Clipboard.js for all .copy-button elements
    const clipboard = new ClipboardJS('.copy-button');
    clipboard.on('success', function (e) {
      e.clearSelection();
      // Briefly change button text to indicate success
      e.trigger.innerText = 'Copied!';
      setTimeout(() => {
        e.trigger.innerHTML = svgIcon + 'Copy'; // restore icon + text
      }, 1500);
    });
    clipboard.on('error', function (e) {
      console.error('Failed to copy', e);
    });
  });
</script>


  <!-- <script>
    document.addEventListener('DOMContentLoaded', function () {
      // 1. Find all <pre><code> blocks
      const codeBlocks = document.querySelectorAll('pre > code');

      codeBlocks.forEach((codeBlock) => {
        const pre = codeBlock.parentNode;

        // 2. Create a container div to hold the code + button
        const container = document.createElement('div');
        container.className = 'code-container';

        // Insert the container before <pre>, then move <pre> inside it
        pre.parentNode.insertBefore(container, pre);
        container.appendChild(pre);

        // 3. Create a "Copy" button
        const button = document.createElement('button');
        button.className = 'copy-button';
        button.type = 'button';
        button.innerText = 'Copy';

        // If the code block doesn’t have an ID, assign one so we can target it
        let codeId = codeBlock.id;
        if (!codeId) {
          codeId = 'codeblock-' + Math.floor(Math.random() * 1000000);
          codeBlock.setAttribute('id', codeId);
        }

        // 4. Point the button at the code block
        button.setAttribute('data-clipboard-target', '#' + codeId);

        // 5. Insert button into the container
        container.insertBefore(button, pre);
      });

      // 6. Initialize Clipboard.js
      const clipboard = new ClipboardJS('.copy-button');
      clipboard.on('success', function (e) {
        e.clearSelection();
        // Change button text to confirm copy
        e.trigger.textContent = 'Copied!';
        setTimeout(() => {
          e.trigger.textContent = 'Copy';
        }, 2000);
      });
      clipboard.on('error', function (e) {
        console.error('Failed to copy', e);
      });
    });
  </script> -->


</footer></body>

</html>

