<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Advanced Tutorial | OrthoFinder</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="Advanced Tutorial" />
<meta name="author" content="David Emms, Jonathan Holmes, Laurence Belcher, Yi Liu" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Advanced tutorial for using OrthoFinder3 OrthoFinder3 provides a new workflow to assign new genes from new species to an already inferred set of orthogroups for a smaller, core group of species. If you are analysing &gt;100 species, we reccommend that you use this scalable implementation. To do this effectively, we need to pick a good set of core species. If we are running OrthoFinder on some bees, some moths, and some flies, we want to build our core orthogroups on a phylogenetically diverse set of those species. If instead we chose randomly and ended up using all moths as our core species, we might end up with less accurate orthogroups, and a longer runtime. If you have a phylogenetic tree, you can use tools like Phylogenetic Diversity Analyzer to pick a representative set (http://www.cibiv.at/software/pda/). Or, you could just pick a diverse set yourself. We recommend that you pick 64 species for your core set. First, run OrthoFinder on the subset of 64 species orthofinder [options] -f &lt;core&gt; Then, add the additional species to the results of the core run orthofinder [options] --assign &lt;additional&gt; --core &lt;Results_Dir&gt; This workflow also makes it really quick and easy to add new species to previous OrthoFinder runs. For example, if your research group works on various species of Angiosperms you might collectively share a core OrthoFinder results folder with a phylogenetically diverse set of species, which individual researchers could easily add any new species to. An important note is that this workflow requires multiple sequence alignments, so unfortunately you cannot use it to add to OrthoFinder2 results that were run with the default -M dendroblast option. Using Outgroups You can make orthogroup inference more accurate by including outgroup species. You just need to make sure that you use the correct N*.tsv file in Phylogenetic_Hierarchical_Orthogroups to look at the orthogroups (use species tree to discover which one)" />
<meta property="og:description" content="Advanced tutorial for using OrthoFinder3 OrthoFinder3 provides a new workflow to assign new genes from new species to an already inferred set of orthogroups for a smaller, core group of species. If you are analysing &gt;100 species, we reccommend that you use this scalable implementation. To do this effectively, we need to pick a good set of core species. If we are running OrthoFinder on some bees, some moths, and some flies, we want to build our core orthogroups on a phylogenetically diverse set of those species. If instead we chose randomly and ended up using all moths as our core species, we might end up with less accurate orthogroups, and a longer runtime. If you have a phylogenetic tree, you can use tools like Phylogenetic Diversity Analyzer to pick a representative set (http://www.cibiv.at/software/pda/). Or, you could just pick a diverse set yourself. We recommend that you pick 64 species for your core set. First, run OrthoFinder on the subset of 64 species orthofinder [options] -f &lt;core&gt; Then, add the additional species to the results of the core run orthofinder [options] --assign &lt;additional&gt; --core &lt;Results_Dir&gt; This workflow also makes it really quick and easy to add new species to previous OrthoFinder runs. For example, if your research group works on various species of Angiosperms you might collectively share a core OrthoFinder results folder with a phylogenetically diverse set of species, which individual researchers could easily add any new species to. An important note is that this workflow requires multiple sequence alignments, so unfortunately you cannot use it to add to OrthoFinder2 results that were run with the default -M dendroblast option. Using Outgroups You can make orthogroup inference more accurate by including outgroup species. You just need to make sure that you use the correct N*.tsv file in Phylogenetic_Hierarchical_Orthogroups to look at the orthogroups (use species tree to discover which one)" />
<link rel="canonical" href="http://localhost:4000/OrthoFinder/tutorials/advanced-tutorial/" />
<meta property="og:url" content="http://localhost:4000/OrthoFinder/tutorials/advanced-tutorial/" />
<meta property="og:site_name" content="OrthoFinder" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2025-05-01T10:02:16+01:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Advanced Tutorial" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","author":{"@type":"Person","name":"David Emms, Jonathan Holmes, Laurence Belcher, Yi Liu"},"dateModified":"2025-05-01T10:02:16+01:00","datePublished":"2025-05-01T10:02:16+01:00","description":"Advanced tutorial for using OrthoFinder3 OrthoFinder3 provides a new workflow to assign new genes from new species to an already inferred set of orthogroups for a smaller, core group of species. If you are analysing &gt;100 species, we reccommend that you use this scalable implementation. To do this effectively, we need to pick a good set of core species. If we are running OrthoFinder on some bees, some moths, and some flies, we want to build our core orthogroups on a phylogenetically diverse set of those species. If instead we chose randomly and ended up using all moths as our core species, we might end up with less accurate orthogroups, and a longer runtime. If you have a phylogenetic tree, you can use tools like Phylogenetic Diversity Analyzer to pick a representative set (http://www.cibiv.at/software/pda/). Or, you could just pick a diverse set yourself. We recommend that you pick 64 species for your core set. First, run OrthoFinder on the subset of 64 species orthofinder [options] -f &lt;core&gt; Then, add the additional species to the results of the core run orthofinder [options] --assign &lt;additional&gt; --core &lt;Results_Dir&gt; This workflow also makes it really quick and easy to add new species to previous OrthoFinder runs. For example, if your research group works on various species of Angiosperms you might collectively share a core OrthoFinder results folder with a phylogenetically diverse set of species, which individual researchers could easily add any new species to. An important note is that this workflow requires multiple sequence alignments, so unfortunately you cannot use it to add to OrthoFinder2 results that were run with the default -M dendroblast option. Using Outgroups You can make orthogroup inference more accurate by including outgroup species. You just need to make sure that you use the correct N*.tsv file in Phylogenetic_Hierarchical_Orthogroups to look at the orthogroups (use species tree to discover which one)","headline":"Advanced Tutorial","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/OrthoFinder/tutorials/advanced-tutorial/"},"url":"http://localhost:4000/OrthoFinder/tutorials/advanced-tutorial/"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/OrthoFinder/assets/css/style.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/OrthoFinder/feed.xml" title="OrthoFinder" /><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
</head>
<body><header class="site-header">

  <div class="wrapper"><!-- <a class="site-title" rel="author" href="/OrthoFinder/%20/">OrthoFinder</a> -->
    <a class="site-title" rel="author" href="/OrthoFinder/">OrthoFinder</a><nav class="site-nav">
      <input type="checkbox" id="nav-trigger" class="nav-trigger" />
      <label for="nav-trigger">
        <span class="menu-icon">
          <svg viewBox="0 0 18 15" width="18px" height="15px">
            <path
              d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z" />
          </svg>
        </span>
      </label>
      <div class="trigger">
        <!--<a class="page-link" href="/OrthoFinder/">Home</a><a class="page-link" href="/OrthoFinder/about/">About</a><a class="page-link" href="/OrthoFinder/download_and_install/">Download &amp; Install</a><a class="page-link" href="/OrthoFinder/tutorial/">Tutorial</a><a class="page-link" href="/OrthoFinder/faq/">FAQ</a>--><a class="page-link" href="/OrthoFinder/">Home</a><a class="page-link" href="/OrthoFinder/about/">About</a><a class="page-link" href="/OrthoFinder/download_and_install/">Download &amp; Install</a><a class="page-link" href="/OrthoFinder/tutorial/">Tutorial</a><a class="page-link" href="/OrthoFinder/faq/">FAQ</a>
        <a class="page-link" href="https://github.com/OrthoFinder/OrthoFinder" target="_blank" title="GitHub Profile">GitHub</a>
        
      </div>
    </nav></div>
  <script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js"></script>
</header><main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Advanced Tutorial</h1>
    <p class="post-meta"><!-- <time class="dt-published" datetime="2025-05-01T10:02:16+01:00" itemprop="datePublished">
        May 1, 2025
      </time> --><!-- <i class="fa fa-refresh" aria-hidden="true"></i> -->
      <time class="dt-modified" datetime="2025-05-07T00:00:00+01:00" itemprop="dateModified">
        Updated May 7, 2025
      </time>
      <meta itemprop="datePublished" content="2025-05-01T10:02:16+01:00"></p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <h2 id="advanced-tutorial-for-using-orthofinder3">Advanced tutorial for using OrthoFinder3</h2>

<p>OrthoFinder3 provides a new workflow to assign new genes from new species to an already 
inferred set of orthogroups for a smaller, core group of species. If you are analysing &gt;100 species, we reccommend that you use this scalable implementation.</p>

<p>To do this effectively, we need to pick a good set of core species. If we are running
OrthoFinder on some bees, some moths, and some flies, we want to build our core 
orthogroups on a phylogenetically diverse set of those species. If instead we chose 
randomly and ended up using all moths as our core species, we might end up with less 
accurate orthogroups, and a longer runtime.</p>

<p>If you have a phylogenetic tree, you can use tools like Phylogenetic Diversity Analyzer to pick a representative set (http://www.cibiv.at/software/pda/). Or, you could just pick a diverse set yourself. We recommend that you pick 64 species for your core set. 
First, run OrthoFinder on the subset of 64 species</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>orthofinder <span class="o">[</span>options] <span class="nt">-f</span> &lt;core&gt;
</code></pre></div></div>

<p>Then, add the additional species to the results of the core run</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>orthofinder <span class="o">[</span>options] <span class="nt">--assign</span> &lt;additional&gt; <span class="nt">--core</span> &lt;Results_Dir&gt;
</code></pre></div></div>
<p>This workflow also makes it really quick and easy to add new species to previous 
OrthoFinder runs. For example, if your research group works on various species of 
Angiosperms you might collectively share a core OrthoFinder results folder with a 
phylogenetically diverse set of species, which individual researchers could easily add any new species to.</p>

<p>An important note is that this workflow requires multiple sequence alignments, so 
unfortunately you cannot use it to add to OrthoFinder2 results that were run with the
default <code class="language-plaintext highlighter-rouge">-M</code> dendroblast option.</p>

<h3 id="using-outgroups">Using Outgroups</h3>

<p>You can make orthogroup inference more accurate by including outgroup species. You 
just need to make sure that you use the correct <code class="language-plaintext highlighter-rouge">N*.tsv</code> file in 
Phylogenetic_Hierarchical_Orthogroups to look at the orthogroups (use species tree to 
discover which one)</p>

  </div><a class="u-url" href="/OrthoFinder/tutorials/advanced-tutorial/" hidden></a>
</article>
<script>
    document.addEventListener("DOMContentLoaded", () => {
        fetch("https://api.github.com/repos/OrthoFinder/OrthoFinder/releases")
            .then(response => response.json())
            .then(data => {
                const assets = data.flatMap(release => release.assets);
                document.querySelectorAll(".download-count").forEach(el => {
                    const assetName = el.getAttribute("data-asset");
                    const match = assets.find(a => a.name === assetName);
                    if (match && match.download_count !== undefined) {
                        el.innerHTML = `<i class="fa fa-download" aria-hidden="true"></i> ${match.download_count} downloads`;
                    } else {
                        el.textContent = "(not found)";
                    }
                });
            })
            .catch(error => {
                console.error("GitHub API fetch error:", error);
                document.querySelectorAll(".download-count").forEach(el => {
                    el.textContent = "(error)";
                });
            });
    });
</script>
        
      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/OrthoFinder/"></data>

  <div class="wrapper">

    <div class="footer-col-wrapper">
      <div class="footer-col">

        <!-- <p class="feed-subscribe">
          <a href="http://localhost:4000/OrthoFinder/feed.xml">
            <svg class="svg-icon orange">
              <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#rss"></use>
            </svg><span>Subscribe</span>
          </a>
        </p> -->

        <!--<ul class="social-media-list"><li>
  <a rel="me" href="" target="_blank" title="GitHub">
    <svg class="svg-icon grey">
      <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
    </svg>
  </a>
</li>


</ul>
-->
         
        <div class="social-icons">
          
          <a href="https://github.com/OrthoFinder/OrthoFinder" target="_blank" title="GitHub">
            <svg class="svg-icon grey">
              <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
            </svg>
            <span>GitHub</span>
          </a>
          
          <!--  -->
        </div>


          <!--
          <ul class="contact-list">

            Link to the team section instead of listing every author
            <li>
              <a href="/OrthoFinder/about/#team">
                Meet&nbsp;the&nbsp;Team
              </a>
            </li>

          </ul> -->

          <ul class="contact-list">
            <li><a href="/OrthoFinder/about/#team">Meet the Team</a></li>
          </ul>


      </div>
      
      <div class="footer-col">
        <!-- <p>This is a software developed by the [Steven Kelly Lab](http://www.stevekellylab.com/).</p> -->
        <p><p>This is a software developed by the <a href="http://www.stevekellylab.com/">Steven Kelly Lab</a>.</p>
</p>
      </div>
    </div>

    <!-- <div class="social-links"><ul class="social-media-list"><li>
  <a rel="me" href="" target="_blank" title="GitHub">
    <svg class="svg-icon grey">
      <use xlink:href="/OrthoFinder/assets/minima-social-icons.svg#github"></use>
    </svg>
  </a>
</li>


</ul>
</div> -->

  </div>


<script>
  document.addEventListener('DOMContentLoaded', function () {
    // 1) Find all <pre><code> blocks
    const codeBlocks = document.querySelectorAll('pre > code');

    codeBlocks.forEach((codeBlock, index) => {
      // The parent <pre> element
      const pre = codeBlock.parentNode;

      // 2) Create a container DIV and wrap <pre> inside it
      const container = document.createElement('div');
      container.className = 'code-container';
      pre.parentNode.insertBefore(container, pre);
      container.appendChild(pre);

      // 3) Create the copy button
      const button = document.createElement('button');
      button.className = 'copy-button';
      button.type = 'button';

      // 4) Insert an inline SVG icon + "Copy" text
      //    We'll build the SVG as a string, then inject it into button.innerHTML
      const svgIcon = `
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="16"
          height="16"
          fill="currentColor"
          class="bi bi-files"
          viewBox="0 0 16 16"
        >
            <path d="M13 0H5a1 1 0 0 0-1 1v1H3a2 2 0 0 0-2
                  2v9.5A1.5 1.5 0 0 0 2.5 15H9v-1H2.5a.5.5
                  0 0 1-.5-.5V4a1 1 0 0 1 1-1h1v7.5A1.5
                  1.5 0 0 0 5.5 12H11v-1H5.5a.5.5 0 0
                  1-.5-.5V1a.5.5 0 0 1 .5-.5H5v3.5A1.5
                  1.5 0 0 0 6.5 5H13v6h1V1a1 1 0 0 0-1-1z"/>
        </svg>
    `;
      

      
      button.innerHTML = svgIcon + 'Copy';

      // 5) Assign an ID to <code> if it doesn't have one
      if (!codeBlock.id) {
        codeBlock.id = `codeblock-${index}`;
      }

      // 6) Link button to the code block by ID
      button.setAttribute('data-clipboard-target', '#' + codeBlock.id);

      // 7) Insert the button into the container
      container.insertBefore(button, pre);
    });

    // 8) Initialize Clipboard.js for all .copy-button elements
    const clipboard = new ClipboardJS('.copy-button');
    clipboard.on('success', function (e) {
      e.clearSelection();
      // Briefly change button text to indicate success
      e.trigger.innerText = 'Copied!';
      setTimeout(() => {
        e.trigger.innerHTML = svgIcon + 'Copy'; // restore icon + text
      }, 1500);
    });
    clipboard.on('error', function (e) {
      console.error('Failed to copy', e);
    });
  });
</script>


  <!-- <script>
    document.addEventListener('DOMContentLoaded', function () {
      // 1. Find all <pre><code> blocks
      const codeBlocks = document.querySelectorAll('pre > code');

      codeBlocks.forEach((codeBlock) => {
        const pre = codeBlock.parentNode;

        // 2. Create a container div to hold the code + button
        const container = document.createElement('div');
        container.className = 'code-container';

        // Insert the container before <pre>, then move <pre> inside it
        pre.parentNode.insertBefore(container, pre);
        container.appendChild(pre);

        // 3. Create a "Copy" button
        const button = document.createElement('button');
        button.className = 'copy-button';
        button.type = 'button';
        button.innerText = 'Copy';

        // If the code block doesnâ€™t have an ID, assign one so we can target it
        let codeId = codeBlock.id;
        if (!codeId) {
          codeId = 'codeblock-' + Math.floor(Math.random() * 1000000);
          codeBlock.setAttribute('id', codeId);
        }

        // 4. Point the button at the code block
        button.setAttribute('data-clipboard-target', '#' + codeId);

        // 5. Insert button into the container
        container.insertBefore(button, pre);
      });

      // 6. Initialize Clipboard.js
      const clipboard = new ClipboardJS('.copy-button');
      clipboard.on('success', function (e) {
        e.clearSelection();
        // Change button text to confirm copy
        e.trigger.textContent = 'Copied!';
        setTimeout(() => {
          e.trigger.textContent = 'Copy';
        }, 2000);
      });
      clipboard.on('error', function (e) {
        console.error('Failed to copy', e);
      });
    });
  </script> -->


</footer></body>

</html>

